<!DOCTYPE html>
<html>
<head>
	<meta charset="UTF-8">
	<title>About The Project 2 </title>
	<style>
		/* Header Styles  */
		.header {
			background-color: #010a0d; /* Black navy for header  */
			padding: 10px;
			text-align: center; /* Center-align the header content */
		}

		.header h1 {
			font-size: 40px; /* Adjust the font size as desired */
			color: #ededf5; /* To change the color of the text "About The Project" in the header to white */
			font-weight: bold; /* Add this  line to make the text bold */
		}

		/* Page Content Styles */
		.content {
			width: 100%;
			padding: 10px;
			box-sizing: border-box;
			border-left: 1px solid green;
			text-align: justify; /* Add this line to justify the text */
			


		}

		.content p {
			font-size: 20px; /* Adjust the font size for paragraphs */
			margin-left: 20px; /* Adjust the margin value for indentation */
			margin-top: 0;
		}

		/* The <hr> element creates a horizontal rule, which is a self-closing tag and doesn't require a closing tag.
          	You can also style the horizontal line using CSS if you want to customize its appearance. For example:*/ 
		hr {
			border: none;
			border-top: 1px solid black;
			margin: 20px 0;
		}

		h2 {
		       
		       font-weight: bold;
		       margin-bottom: 0;
		}
		
	</style>
</head>
<body>
	<div class="header">
		<h1>About The Project 2 </h1>
	</div>

	<div class="content">
		<h1 style="text-align: center; font-size: 30px; font-weight: bold;">Project 2: Differential Expression Analysis for Biomarker Discovery in Breast Cancer</h1>
		<hr>

<h2 style="font-size: 22px; font-weight: bold;">Background: </h2>
		<h2 style="font-size: 22px; font-weight: bold;">Background: </h2>
		<p>  <span style="font-size: 20px;">Breast cancer is a complex and heterogeneous disease with varying levels of aggressiveness. Accurate classification of breast cancer subtypes and identification of robust biomarkers are essential for personalized treatment decisions and prognostic assessments. In this project, we aimed to conduct a comprehensive analysis of gene expression data to identify differentially expressed genes (DEGs) associated with breast cancer aggressiveness. The ultimate goal was to discover potential biomarkers that could be utilized in the industry for improved diagnosis and treatment strategies.</span> </p>
		

		<h2 style="font-size: 22px; font-weight: bold;">Objectives: </h2>
		<p>  <span style="font-size: 20px; ">The primary objectives of Project 2 were as follows:</span> </p>
		<h2 style="font-size: 20px; text-indent: 25px;">1. Assess the differential gene expression between non-aggressive (Grade 1) and aggressive (Grade 3) breast cancer samples.</h2>
		<h2 style="font-size: 20px; text-indent: 25px;">2. Identify DEGs that exhibit significant expression differences associated with disease aggressiveness. </h2>
		<h2 style="font-size: 20px; text-indent: 25px;">3. Utilize bioinformatics tools and R packages to facilitate data analysis and visualization.  </h2>
		<h2 style="font-size: 20px; text-indent: 25px;">4. Provide insights into potential biomarkers for industry application in breast cancer diagnosis and treatment. </h2>


		<h2 style="font-size: 22px; font-weight: bold;">Methodology: </h2>
		<h2 style="font-size: 20px; text-indent: 25px;">1.	Data Acquisition and Preprocessing: </h2>
		<p style="font-size: 20px; text-indent: 35px;">The gene expression datasets, GSE25055, GSE7390, and GSE11121, were retrieved from the Gene Expression Omnibus (GEO) database using the GEOquery R package.</p>
		<p style="font-size: 20px; text-indent: 35px;">Quality control measures were implemented to ensure data integrity, including assessment of sample quality and data normalization.</p>
		<h2 style="font-size: 20px; text-indent: 25px;">2.	Differential Expression Analysis:</h2>
		<p style="font-size: 20px; text-indent: 35px;">The limma R package was employed for differential expression analysis. This package is widely used for its robust statistical methods and flexibility in handling microarray data.</p>
		<p style="font-size: 20px; text-indent: 35px;">Individual dataset analyses were performed using the limma package to identify DEGs specific to each dataset.</p>
		<p style="font-size: 20px; text-indent: 35px;">Fold change (FC) and false discovery rate (FDR) values were calculated to assess the magnitude and statistical significance of gene expression changes.</p>
		<h2 style="font-size: 20px; text-indent: 25px;">3.	Integration and Visualization: </h2>
		<p style="font-size: 20px; text-indent: 35px;">The identified DEGs from each dataset were further analyzed to obtain a common set of DEGs across the three datasets.</p>
		<p style="font-size: 20px; text-indent: 35px;">The ggvenn package was used to visualize the overlapping DEGs among the datasets, providing a visual representation of shared and unique genes.</p>
		<p style="font-size: 20px; text-indent: 35px;">The tidyverse package, along with plotly, was utilized for data manipulation, exploration, and interactive visualizations, facilitating the interpretation of the results.</p>
		

		<h2 style="font-size: 22px; font-weight: bold;">Results: </h2>
		<p>  <span style="font-size: 20px;">The differential expression analysis revealed a significant number of DEGs associated with breast cancer aggressiveness. The limma package enabled the identification of specific gene expression patterns between non-aggressive and aggressive tumor groups. Through integration and comparison of the individual dataset analyses, a common set of 77 DEGs was identified across the three datasets. The ggvenn package aided in visualizing the overlap of DEGs, highlighting both shared and distinct gene signatures.</span> </p>


		<h2 style="font-size: 22px; font-weight: bold;">Discussion: </h2>
		<p>  <span style="font-size: 20px;">The utilization of the limma package allowed us to uncover important DEGs associated with breast cancer aggressiveness. By analyzing individual datasets, we were able to identify gene expression differences specific to each dataset, providing valuable insights into the molecular mechanisms underlying disease progression. The integration of multiple datasets enhanced the robustness of the findings and provided a more comprehensive view of DEGs associated with breast cancer aggressiveness.</span> </p>
		<p>  <span style="font-size: 20px;">The use of R packages such as GEOquery, limma, tidyverse, plotly, and ggvenn was instrumental in facilitating data acquisition, preprocessing, differential expression analysis, and result visualization. GEOquery allowed seamless retrieval of gene expression datasets from the GEO database, while limma provided powerful statistical methods for detecting DEGs. The tidyverse package and plotly allowed for efficient data manipulation, exploration, and interactive visualizations, aiding in the interpretation and communication of the results. The ggvenn package facilitated the comparison and visualization of overlapping DEGs, enabling a deeper understanding of shared and unique gene signatures.</span> </p>
		<p>  <span style="font-size: 20px;">The results of this project provide valuable insights into potential biomarkers associated with breast cancer aggressiveness, laying the foundation for further validation and potential industry applications. The identification of DEGs through comprehensive differential expression analysis sets the stage for future investigations and the development of personalized diagnostic and therapeutic strategies for breast cancer patients.</span> </p>


	
		
	
	</div>

</body>
</html>
