<!DOCTYPE html>
<html>
<head>
	<meta charset="UTF-8">
	<title>About</title>
	<style>
		/* Header Styles  */
		.header {
			background-color: #010a0d; /* Black navy for header  */
			padding: 10px;
			text-align: center; /* Center-align the header content */
		}

		.header h1 {
			font-size: 40px; /* Adjust the font size as desired */
			color: #ededf5; /* To change the color of the text "About The Project" in the header to white */
			font-weight: bold; /* Add this  line to make the text bold */
		}

		/* Page Content Styles */
		.content {
			width: 100%;
			padding: 10px;
			box-sizing: border-box;
			border-left: 1px solid green;
			text-align: justify; /* Add this line to justify the text */
			


		}

		.content p {
			font-size: 20px; /* Adjust the font size for paragraphs */
			margin-left: 20px; /* Adjust the margin value for indentation */
			margin-top: 0;
		}

		/* The <hr> element creates a horizontal rule, which is a self-closing tag and doesn't require a closing tag.
          	You can also style the horizontal line using CSS if you want to customize its appearance. For example:*/ 
		hr {
			border: none;
			border-top: 1px solid black;
			margin: 20px 0;
		}
	</style>
</head>
<body>
	<div class="header">
		<h1>About </h1>
	</div>

	<div class="content">
		<p>
		<strong> These three code snippets represent different approaches and analyses applied to the datasets, including single dataset analysis, differential expression analysis with multiple datasets, and meta-analysis to combine the results from multiple datasets. </strong>
		</p>
		
		<h2>1. R CODE (1 Dataset)</h2>
		<p>
		In this code, a differential expression analysis was performed on the breast cancer dataset GSE25055 using the limma package in R. The analysis aimed to compare the effectiveness of molecular subtypes and the grading system in distinguishing aggressive and non-aggressive forms of breast cancer. By identifying genes with significant differential expression and visualizing the results through volcano plots, the code explored the potential of these classifications as indicators of aggressiveness.
		</p>
		<hr>

		<h2>2. R CODE (4 Dataset analysis - FC & FDR)</h2>
		<p>
		In this code, a differential expression analysis is conducted using four datasets with the same platform. The analysis utilizes the limma package for differential expression and the tidyverse package for data manipulation. The objective of this analysis is to compare the gene expression profiles between two grades of cancer (Grade 1 and Grade 3). Statistical measures such as fold change (FC) and false discovery rate (FDR) are used to identify genes that are differentially expressed between the two grades. Additionally, the code incorporates the ggvenn package to generate Venn diagrams, aiming to identify the common differentially expressed genes (DEGs) across the four datasets. The volcano plots are created to visualize the results, and the final outcome is the identification of DEGs as well as the common DEGs shared among the datasets.
		</p>
		<hr>

		<h2>3. R CODE (4 Dataset Meta-analysis - Z score & FDR)</h2>
		<p>
		This code snippet involves performing a meta-analysis using the Genemeta package. The analysis integrates the same four datasets from the previous code snippet. The effect size, represented by the z score, and the false discovery rate (FDR) are utilized for the meta-analysis. The objective is to identify and obtain the meta-combined differentially expressed genes (DEGs) across the four datasets.
		</p>
		<hr>

		<h1 style="text-align: center; font-size: 30px; font-weight: bold;">Project Summary</h1>

		<h2>1-	The Heterogeneity and Aggressiveness of Breast Cancer:</h2>
		<p>
		The nature of breast cancer is characterized by heterogeneity, resulting in different levels of aggressiveness. Given the heterogeneity and varying aggressiveness levels of breast cancer, precise classification is imperative for guiding treatment decisions and accurately forecasting patient outcomes. This project aimed to differentiate aggressiveness levels in breast cancer by employing two classification approaches: the grade-based classification (comparing grade 1 (non-aggressive) to grade 3 (aggressive)) and the molecular subtype-based classification (comparing Luminal A, Normal-like (non-aggressive) to Luminal B, Basal-like, HER2-positive (aggressive) subtypes). The study consisted of three interconnected projects, referred to as Project 1, 2, and 3.
		</p>


		<h2>2-	Objectives and Approach of Project 1: Assessing Grade-based and Subtype-based Classification:</h2>
		<p>
		Project 1 focused on assessing the effectiveness of the grade-based and subtype-based classification approaches in differentiating aggressiveness levels in breast cancer. The dataset GSE25055, obtained from pre-treatment tumor biopsies, included gene expression data for both grading system and molecular subtypes. It was subjected to a comprehensive differential expression analysis using the limma package in R. Significant gene expression differences were identified between non-aggressive and aggressive tumor groups based on both classification approaches, highlighting their potential for accurate differentiation of aggressiveness levels. 
		</p>


		<h2>3-	The Shift in Focus: Future Directions for Subtype-based Analysis:</h2>
		<p>
		Nevertheless, due to the larger number of samples with grading information compared to those with subtype information, the project prioritized grade-based classification for project 2 and 3. However, in the future, the research will shift its focus to conducting a differential expression (DE) analysis using subtype-based classification. This analysis will allow for a comparison of the results obtained from the subtype-based analysis with the existing findings derived from the grade-based classification, providing a more comprehensive understanding of disease progression and facilitating the discovery of novel insights.
		</p>


		<h2>4-	Project 2: Analyzing Individual Datasets and Comparing Gene Expression Profiles:</h2>
		<p>
		Building upon the findings of Project 1, Project 2 focuses on analyzing individual datasets and comparing the gene expression profiles between non-aggressive (Grade 1) and aggressive (Grade 3) breast cancer samples using microarray datasets. Three microarray datasets (GSE25055, GSE7390, GSE11121) were selected, and differential expression analysis was performed using the limma package. The analysis revealed differentially expressed genes (DEGs) associated with the aggressiveness of breast cancer, providing molecular signatures indicative of disease severity.
		</p>


		<h2>5-	Project 3: Meta-analysis for Robust Biomarker Discovery:</h2>
		<p>
		In Project 3, the focus was on utilizing a meta-analysis approach to integrate multiple breast cancer microarray datasets (GSE25055, GSE7390, GSE11121) with the goal of robust biomarker discovery. The main objective of Project 3 was to generate a more extensive set of differentially expressed genes (DEGs) and enhance the overall comprehensiveness of the findings. Comprehensive data preprocessing, including quality control, normalization, and batch effect correction, was performed. Meta-analysis was conducted using the random-effects model, combining DEG results from individual dataset analyses. Functional enrichment analysis was performed on the identified DEGs, providing insights into the underlying biological processes associated with breast cancer aggressiveness.
		</p>


		<h2>6-	Comparison of Objectives: Project 2 vs. Project 3:</h2>
		<p>
		Project 2 involved the analysis of three breast cancer microarray datasets (GSE25055, GSE7390, GSE11121) using the limma package. The primary objective was to identify differentially expressed genes (DEGs) specific to each dataset and compare gene expression profiles between non-aggressive (Grade 1) and aggressive (Grade 3) breast cancer samples. The aim was to gain insights into the molecular differences associated with varying levels of aggressiveness using individual dataset analyses.
		</p>

		<p>
		In contrast, Project 3 focused on conducting a meta-analysis to integrate the results from the three aforementioned datasets. The objective was to obtain a more comprehensive understanding of gene expression changes and biomarker significance. By applying the GeneMeta package and employing the random-effects model (REM), Project 3 aimed to identify common DEGs across the datasets, calculate Z scores and false discovery rate (FDR) values, and generate a larger set of meta-combined DEGs. The meta-analysis approach enhanced the robustness and reliability of the findings by incorporating data from multiple sources.
		</p>

		<p>
		In summary, while Project 2 aimed to explore DEGs within each dataset separately, Project 3 utilized a meta-analysis strategy to combine the results from multiple datasets. The meta-analysis in Project 3 provided a broader perspective and more comprehensive insights into gene expression changes and potential biomarkers associated with breast cancer aggressiveness.
		</p>

		<p>
		Overall, this project aimed to advance the understanding of breast cancer classification and identification of robust indicators of aggressiveness. The integration of different classification approaches and the discovery of DEGs through comprehensive analysis contributed to the growing knowledge in the field. The results obtained from these projects lay the foundation for further investigations, enabling personalized and targeted treatments for breast cancer patients.
		</p>
		


		<p>
		</p>

		

	
	</div>

</body>
</html>
