<!DOCTYPE html>
<html>
<head>
	<meta charset="UTF-8">
	<title>About The Project 3 </title>
	<style>
		/* Header Styles  */
		.header {
			background-color: #010a0d; /* Black navy for header  */
			padding: 10px;
			text-align: center; /* Center-align the header content */
		}

		.header h1 {
			font-size: 40px; /* Adjust the font size as desired */
			color: #ededf5; /* To change the color of the text "About The Project" in the header to white */
			font-weight: bold; /* Add this  line to make the text bold */
		}

		/* Page Content Styles */
		.content {
			width: 100%;
			padding: 10px;
			box-sizing: border-box;
			border-left: 1px solid green;
			text-align: justify; /* Add this line to justify the text */
			


		}

		.content p {
			font-size: 20px; /* Adjust the font size for paragraphs */
			margin-left: 20px; /* Adjust the margin value for indentation */
			margin-top: 0;
		}

		/* The <hr> element creates a horizontal rule, which is a self-closing tag and doesn't require a closing tag.
          	You can also style the horizontal line using CSS if you want to customize its appearance. For example:*/ 
		hr {
			border: none;
			border-top: 1px solid black;
			margin: 20px 0;
		}
	</style>
</head>
<body>
	<div class="header">
		<h1>About The Project 3 </h1>
	</div>

	<div class="content">
		
		<h1 style="text-align: center; font-size: 30px; font-weight: bold;">Title: Integration of Gene Expression Data through Meta-Analysis for Robust Biomarker Discovery in Breast Cancer</h1>
		<hr>

		
		<h2 style="font-size: 22px; font-weight: bold;">Background: </h2>
		<p>  <span style="font-size: 20px; margin-bottom: 0;">Given the intricate nature of breast cancer, which exhibits substantial heterogeneity, the identification of reliable biomarkers assumes paramount importance, allowing for refined diagnostic modalities and more effective treatment strategies. Project 3 aimed to leverage the power of meta-analysis to integrate multiple breast cancer microarray datasets (GSE25055, GSE7390, GSE11121) for robust biomarker discovery. The objective was to identify a more extensive set of differentially expressed genes (DEGs) and gain a comprehensive understanding of gene expression changes associated with breast cancer aggressiveness. This analysis was conducted with a focus on industry applications for biomarker discovery.</span> </p>


		<h2 style="font-size: 22px; font-weight: bold;">Objectives:</h2>
		<p>  <span style="font-size: 20px; margin-bottom: 0;">The main objective of Project 3 was to perform a meta-analysis of the selected breast cancer microarray datasets to identify robust DEGs associated with breast cancer. The specific objectives were:</span> </p>
		<h2 style="font-size: 20px; text-indent: 25px; font-weight: normal;">1.	Integrate the gene expression data from multiple datasets to reduce dataset-specific biases.</h2>
		<h2 style="font-size: 20px; text-indent: 25px; font-weight: normal;">2.	Identify common DEGs across the datasets using the random-effects model (REM) approach.</h2>
		<h2 style="font-size: 20px; text-indent: 25px; font-weight: normal;">3.	Calculate Z scores and false discovery rate (FDR) values for the meta-combined DEGs.</h2>
		<h2 style="font-size: 20px; text-indent: 25px; font-weight: normal;">4.	Conduct functional enrichment analysis to gain insights into the biological processes associated with breast cancer aggressiveness.</h2>

		<h2 style="font-size: 22px; font-weight: bold;">Methodology:</h2>
		<h2 style="font-size: 20px; text-indent: 25px; font-weight: normal;">1.	Dataset Selection: Three microarray datasets (GSE25055, GSE7390, GSE11121) were chosen for meta-analysis. These datasets provided gene expression profiles of breast cancer samples.</h2>
		<h2 style="font-size: 20px; text-indent: 25px; font-weight: normal;">2.	Data Preprocessing: Comprehensive data preprocessing steps were performed, including quality control, normalization, and batch effect correction. The GEOquery package was used to retrieve the data, while the tidyverse, reshape2, and vsn packages were employed for data preprocessing.</h2>
		<h2 style="font-size: 20px; text-indent: 25px; font-weight: normal;">3.	Visualization and Assessment of Data: Box plots, principal component analysis (PCA) plots, density plots, and multidimensional scaling (MDS) plots were generated to visualize the effects of normalization and batch effect correction on the datasets. This allowed for the assessment of data quality and reduction of technical variations.</h2>
		<h2 style="font-size: 20px; text-indent: 25px; font-weight: normal;">4.	Meta-Analysis Method Selection: The presence of heterogeneity among the datasets was assessed using the Cochran Q Test and QQ plot analysis. Based on the results, the random-effects model (REM) was selected as the appropriate approach for the meta-analysis.</h2>
		<h2 style="font-size: 20px; text-indent: 25px; font-weight: normal;">5.	Meta-Analysis and DEG Identification: The GeneMeta package was utilized for the meta-analysis, combining the DEG results obtained from the individual dataset analyses. This allowed for the identification of common DEGs across the datasets. Z scores and false discovery rate (FDR) values were calculated to assess the significance of the DEGs.</h2>
		<h2 style="font-size: 20px; text-indent: 25px; font-weight: normal;">For the REM meta-analysis in Project 3, the GeneMeta package was utilized to calculate Z scores for the meta-combined differentially expressed genes (DEGs). The Z scores serve as a measure of the significance and directionality of gene expression changes across the multiple datasets. A positive Z score indicates an upregulation of gene expression, while a negative Z score indicates a downregulation. The Z scores take into account the combined effect sizes and their variability across the datasets, providing a standardized metric to assess the magnitude of gene expression differences in the context of the meta-analysis. Additionally, the Z scores can be used to prioritize genes and identify potential biomarkers associated with breast cancer aggressiveness.</h2>
		<h2 style="font-size: 20px; text-indent: 25px; font-weight: normal;">6.	Functional Analysis: The identified DEGs from the meta-analysis were subjected to functional enrichment analysis. Various packages such as msigdbr, MSigDB, fgsea, and PANTHER were used for gene ontology (GO) analysis, KEGG pathway analysis, gene set enrichment analysis (GSEA), and protein-protein interaction (PPI) network analysis. This provided insights into the underlying biological processes associated with breast cancer aggressiveness.</h2>

		<h2 style="font-size: 22px; font-weight: bold;">Results:</h2>
		<p>  <span style="font-size: 20px; margin-bottom: 0;">The meta-analysis approach successfully integrated the gene expression data from multiple datasets, reducing dataset-specific biases. The analysis identified a larger set of DEGs compared to individual dataset analyses. The Z scores and FDR values provided statistical measures of significance for the meta-combined DEGs. The functional enrichment analysis revealed key biological processes and pathways associated with breast cancer aggressiveness.</span> </p>

		<h2 style="font-size: 22px; font-weight: bold;">Discussion:</h2>
		<p>  <span style="font-size: 20px; margin-bottom: 0;">The meta-analysis approach employed in Project 3 enhanced the robustness and reliability of the findings by integrating data from multiple sources. This comprehensive analysis allowed for the identification of a more extensive set of DEGs, providing a broader perspective on gene expression changes in breast cancer. The functional enrichment analysis provided valuable insights into the underlying biological processes associated with breast cancer aggressiveness, facilitating the discovery of potential biomarkers for industry applications.</span> </p>
		<p>  <span style="font-size: 20px; margin-bottom: 0;">The R packages used in Project 3, including tidyverse, GEOquery, reshape2, caret, GeneMeta, rgl, sva, plotly, ggvenn, ggpubr, ComplexHeatmap, RColorBrewer, msigdbr, MSigDB, fgsea, locfit, and vsn, were carefully selected to ensure efficient data retrieval, preprocessing, meta-analysis, visualization, and functional analysis. These packages offered a wide range of functionalities and tools for comprehensive analysis of the breast cancer microarray datasets.</span> </p>
		<p>  <span style="font-size: 20px; margin-bottom: 0;">Overall, Project 3 contributed to the advancement of biomarker discovery in breast cancer by employing a meta-analysis approach. The integrated analysis of multiple datasets provided a more comprehensive understanding of gene expression changes and identified robust DEGs associated with breast cancer. The findings of this project hold significant potential for industry applications in personalized medicine and targeted therapies for breast cancer patients.</span> </p>


	
	</div>

</body>
</html>
