<!DOCTYPE html>
<html>
<head>
	<meta charset="UTF-8">
	<title>R Code For One Dataset</title>
	<style>
		/* CSS Styles */
		/* To highlight different  parts of the text in red, green, and blue, you can use the <span> element and apply CSS styles to  it */ 
		
		.orange-text {
			color: orange;
			font-weight: bold;
		}
		.dark-yellow-text {
			color: hsl(50, 100%, 30%); /* Adjust the hue and saturation values as needed */
			font-weight: bold;
		}
		.brown-text {
			color: hsl(30, 100%, 30%); /* Adjust the hue and saturation values as needed */
			font-weight: bold;
		}
		.green-text {
			color: green;
			font-weight: bold;
		}
		.purple-text {
			color: purple;
			font-weight: bold;
		}
		.blue-text {
			color: blue;
			font-weight: bold;
		}

		/* Header Styles */
		header {
			background-color: #274f4f;
			color: white;
			text-align: center;
			padding: 5px;
		}

		h1 {
			color: white; /* Adjust the font color as desired */
			font-size: 30px; /* Adjust the font size as desired */
			font-weight: bold; /* Add font weight as desired */
			text-align: center; /* Center the header title */
			padding: 0px; /* Add padding around the header title */
		}
		
		hr {
			width: 100%; /* Adjust the width as desired */
		}
		
		/* Drop-down Styles */
		.drop-down-content {
			display: none;
		}

		.dropdown-trigger {
			cursor: pointer;
		}

		.drop-down-content.active {
			display: block;
		}
		
		
		h3.dropdown-trigger {
		font-size: 24px; /* Adjust the font size as desired */
		}

	        /* Adjust font size for paragraph under 1-1- Explanation:*/
		.drop-down-content p {
			font-size: 16px; /* Adjust the font size as desired */
		}

		/* Adjust font size for list items */
		.drop-down-content ul li {
			font-size: 16px; /* Adjust the font size as desired */
		}

			/* Adjust font size for the explanation paragraph */
		#explanation-content p {
			font-size: 20px; /* Adjust the font size as desired */
		}
		
		/* Adjust font size for the headings */
		#plots-tables-content ul li a {
			font-size: 20px; /* Adjust the font size as desired */
		}
		
		/* Adjust font size for the code block */
		#code-content pre code {
			font-size: 17px; /*  Adjust the font size as desired */
		}
		
		 /* Footer Styles */
		.footer {
		        background-color: #333;
		        color: #fff;
		        padding: 10px;
		        text-align: center;
		        position: fixed;
		        left: 0;
		        bottom: 0;
		        width: 100%;
		}
		
		  
		
	</style>
	
</head>
<body>
	<header>
		<h1>R Code <br> (One Dataset)</h1>
	</header>
	
	
	<h2 style="font-size: 28px; text-align: center;">1- Differential Expression Analysis of Breast Cancer Dataset (GSE25055) using limma in R: Comparing Molecular Subtypes and Grading System</h2>

	
	<hr>
	<h3 class="dropdown-trigger" onclick="toggleDropDown('explanation-content')">About this code</h3>
	<div class="drop-down-content" id="explanation-content" style="text-align: justify;">
		<p> In this code, we performed a differential expression analysis on the breast cancer dataset GSE25055 using the limma package in R. The dataset contains gene expression data for different grades of breast cancer (Grade 1 and Grade 3) as well as molecular subtypes (Luminal A, Normal-like, Luminal B, Basal-like, and HER2). </p> 

		<p> The purpose of this analysis was to compare the ability of two classifications, molecular subtypes and the grading system, to differentiate between different levels of aggressiveness in breast cancer. Based on a literature review, we hypothesized that Luminal A and Normal-like molecular subtypes, as well as Grade 1, represent non-aggressiveness, while Luminal B, Basal-like, HER2 molecular subtypes, and Grade 3 represent aggressiveness. </p> 
		<p> To perform the analysis, we used the GEOquery package to download the GSE25055 dataset from the GEO database. We also utilized the tidyverse package for data manipulation and preprocessing, and the plotly package for creating interactive volcano plots. </p> 
		<p> The main goal of the analysis was to identify genes that showed significant differential expression between the non-aggressive and aggressive groups based on the two classifications. We used the limma package to apply appropriate statistical tests and adjust for multiple testing. The volcano plots were generated to visualize the statistical significance (p-value) and fold change of differentially expressed genes. </p> 
		<p> By comparing the results of the molecular subtypes and grading system analyses, we aimed to determine which classification was better in terms of statistically significant and biologically meaningful differences between the non-aggressive and aggressive groups. </p> 
		<p> Overall, this code provides insights into the differential expression analysis of breast cancer dataset GSE25055 and investigates the potential of molecular subtypes and the grading system as indicators of aggressiveness in breast cancer. </p>

	</div>
	<hr>

	
	<h3 class="dropdown-trigger" onclick="toggleDropDown('plots-tables-content')">Generated Plots & Tables:</h3>
	<div class="drop-down-content" id="plots-tables-content">
		<ul>
			<li><a href="#">Volcano plots for comparing grade 1 vs grade 3 (GSE25055 dataset)</a></li>
			<li><a href="#">Volcano plots for comparing Luminal A, Normal-like vs Luminal B, Basal-like, and HER2 (GSE25055 dataset)</a></li>
			<li><a href="#">Tables</a></li>
		</ul>
	</div>
	<hr>

	<h3 class="dropdown-trigger" onclick="toggleDropDown('code-content')">Code</h3>
	<div class="drop-down-content" id="code-content">
		<pre><code>
			<!--
			<span class="green-text">GREEN TEXTTTTTT</span> 
			<span class="purple-text">PURPLE TEXTTTTTT</span> 
			<span class="orange-text">ORANGE TEXTTTT</span> 
			<span class="dark-yellow-text">DARK YELLOW TEXTTT</span> 
			<span class="brown-text">BROWN TEXTTTT</span> 
			<span class="blue-text">BLUE TEXTTTTT</span> 
			<strong>BOLD TEXTTTTTT</strong> 
			-->
<!-- put the R CODE comments in this format:    <span style="color: #04cc04; font-style: italic;"></span> -->
<span style="color: #c910c9; font-style: italic;">require</span>(limma) 
<span style="color: #c910c9; font-style: italic;">require</span>(tidyverse)
<span style="color: #c910c9; font-style: italic;">require</span>(plotly)


 
DownloadGEO <- <span style="color: #c910c9; font-style: italic;">function</span>(GSEname) {
  <span style="color: #04cc04; font-style: italic;"># Create a folder</span>
  dir.create(GSEname, showWarnings = FALSE)
  
  <span style="color: #04cc04; font-style: italic;"># Increase memory</span>
  Sys.setenv("VROOM_CONNECTION_SIZE" = 131072 * 10)
  
  <span style="color: #04cc04; font-style: italic;"># Download dataset</span>
  gset <- getGEO(
    GSEname,
    GSEMatrix = TRUE,
    AnnotGPL = TRUE,
    destdir = paste0(getwd(), "/", GSEname)
  )
  
  <span style="color: #c910c9; font-style: italic;">if</span>(!is.null(gset)) {
    <span style="color: #04cc04; font-style: italic;"># Extracting matrix, phenotype, and annotation</span>
    matrix <- gset[[paste0(GSEname, "_series_matrix.txt.gz")]]@assayData[["exprs"]]
    matrix <- cbind(rownames(matrix), matrix)
    colnames(matrix)[1] <- "ID"
    
    phenotype <- gset[[paste0(GSEname, "_series_matrix.txt.gz")]]@phenoData@data
    phenotype <- cbind(rownames(phenotype), phenotype)
    colnames(phenotype)[1] <- "ID"
    
    annot <- gset[[paste0(GSEname, "_series_matrix.txt.gz")]]@featureData@data
    
    <span style="color: #04cc04; font-style: italic;"># Writing the extracted variables</span>
    write.csv(matrix, paste0(GSEname, "/", "matrix.csv"), quote = FALSE, row.names = FALSE)
    write_tsv(phenotype, paste0(GSEname, "/", "phenotype.tsv"))
    write_tsv(annot, paste0(GSEname, "/", "annot.tsv"))
    
    <span style="color: #04cc04; font-style: italic;"># Removing variables</span>
    rm(gset, matrix, phenotype, annot)
    
    message("+++++ Dataset is ready to analyze. +++++")
  }
}


DEGanalysis <- <span style="color: #c910c9; font-style: italic;">function</span>(projname,
                        compare = "subtype",
                        groups,
                        adjust = "fdr") {
  <span style="color: #04cc04; font-style: italic;"># Reading the files</span>
  matrix <- read_csv(paste0(projname, "/", "matrix.csv"))
  matrix <- data.frame(matrix)
  rownames(matrix) <- matrix[,1]
  matrix <- matrix[,-1]
  phenotype <- read_tsv(paste0(projname, "/", "phenotype.tsv"))
  phenotype <- data.frame(phenotype)
  rownames(phenotype) <- phenotype[,1]
  phenotype <- phenotype[,-1]
  annot <- read_tsv(paste0(projname, "/", "annot.tsv"))
  annot <- data.frame(annot)
  rownames(annot) <- annot[,1]
  annot <- annot[,-1]
  
  <span style="color: #04cc04; font-style: italic;"># Preparing the breast cancer subtype into a dataframe</span>
  phecoln <- grep("pam50", phenotype)
  pheno <- data.frame(phenotype[ ,phecoln])
  colnames(pheno) <- "subtype"
  pheno$subtype <- gsub("pam50.+:", "", pheno$subtype)
  rownames(pheno) <- rownames(phenotype)

  <span style="color: #04cc04; font-style: italic;"># Preparing the breast cancer grades into a dataframe</span>
  gradecoln <- grep("grade:", phenotype)
  grade <- data.frame(phenotype[ ,gradecoln])
  colnames(grade) <- "grade"
  grade$grade <- gsub("grade:", "", grade$grade)
  grade$grade <- gsub("=.+", "", grade$grade)
  rownames(grade) <- rownames(phenotype)
  
  <span style="color: #04cc04; font-style: italic;"># log2 transformation if the matrix is not already transformed</span>
  qx <- as.numeric(quantile(matrix, c(0., 0.25, 0.5, 0.75, 0.99, 1.0), na.rm=T))
  LogC <- (qx[5] > 100) || (qx[6]-qx[1] > 50 && qx[2] > 0)
  <span style="color: #c910c9; font-style: italic;">if</span>(LogC){
    matrix[which(matrix <= 0)] <- NaN
    matrix <- log2(matrix)
  }

  <span style="color: #04cc04; font-style: italic;"># Extracting the sample ID according to the groups names specified by user</span>
  <span style="color: #c910c9; font-style: italic;">if</span>(compare == "subtype"){
    grouplis <- list() 
    <span style="color: #c910c9; font-style: italic;">for</span>(i <span style="color: #c910c9; font-style: italic;">in</span> seq_along(groups)){
      name <- paste0(groups[[i]], collapse = "|")
      rownum <- grep(name, pheno$subtype)
      grouplis[[names(groups)[i]]] <- rownames(pheno)[rownum]
    }
  } <span style="color: #c910c9; font-style: italic;">else</span>{
    grouplis <- list()
    <span style="color: #c910c9; font-style: italic;">for</span>(i <span style="color: #c910c9; font-style: italic;">in</span> seq_along(groups)){
      name <- paste0(groups[[i]], collapse = "|")
      rownum <- grep(name, grade$grade)
      grouplis[[names(groups)[i]]] <- rownames(grade)[rownum]
    }
  }
  
  <span style="color: #04cc04; font-style: italic;"># Renaming the matrix file according to the group names specified by user</span> 
  <span style="color: #c910c9; font-style: italic;">for</span>(i <span style="color: #c910c9; font-style: italic;">in</span> seq_along(grouplis)){
    matname <- paste0(grouplis[[i]], collapse = "|")
    matcoln <- grep(matname, colnames(matrix))
    colnames(matrix)[matcoln] <- paste0(names(grouplis)[i], 1:length(grouplis[[i]]))
  }
  
  <span style="color: #04cc04; font-style: italic;"># Keeping only samples that are renamed and order them</span>
  matnames <- grep(paste0(names(grouplis), collapse = "|"), colnames(matrix))
  matrix <- matrix[, matnames]
  matrix <- matrix[, order(colnames(matrix))]
  
  <span style="color: #04cc04; font-style: italic;"># Creating a factor of 0 and 1 according to the length of groups specified by user</span>
  gname <- list()
  <span style="color: #c910c9; font-style: italic;">for</span>(i <span style="color: #c910c9; font-style: italic;">in</span> seq_along(grouplis)){
    number <- length(grep(names(grouplis)[i], colnames(matrix)))
    gname[[i]] <- factor(rep(i - 1, number))
  }
  grouping <- unlist(gname)
  
  <span style="color: #04cc04; font-style: italic;"># Making the design matrix</span>
  design <- model.matrix(~ 0 + grouping)
  colnames(design) <- names(grouplis)
  
  <span style="color: #04cc04; font-style: italic;"># Fitting the linear model</span>
  fit1 <- lmFit(matrix, design)
  l <- length(names(grouplis))
  cts <- paste0(names(grouplis)[l:1], collapse = "-")
  cont.matrix <- makeContrasts(contrasts = cts, levels = design)
  fit2 <- contrasts.fit(fit1, cont.matrix)
  fit2 <- eBayes(fit2)
  
  <span style="color: #04cc04; font-style: italic;"># Creating top table with all genes</span>
  DEGs <- topTable(fit2, adjust = adjust, number = Inf)
  DEGs <- cbind(rownames(DEGs), DEGs)
  colnames(DEGs)[1] <- "ID" 
  
  <span style="color: #04cc04; font-style: italic;"># Preparing the annotation variable</span>
  annotation <- annot[, c("Gene.symbol", "Chromosome.location", "GO.Function")]
  annotation <- cbind(rownames(annotation), annotation)
  colnames(annotation)[1] <- "ID" 
  
  <span style="color: #04cc04; font-style: italic;"># Making sure that both DEGs and annotation row names are the same</span>
  annotation <- annotation[order(annotation$ID), ]
  DEGs <- DEGs[order(DEGs$ID), ]
  
  <span style="color: #04cc04; font-style: italic;"># Adding annotation to the top table</span>
  DEGs <- cbind(DEGs,  annotation[, c(2, 3, 4)])
  
  <span style="color: #04cc04; font-style: italic;"># Removing missing gene symbols</span>
  genemissing <- which(is.na(DEGs$Gene.symbol) == TRUE)
  DEGs <- DEGs[-genemissing, ]
  
  <span style="color: #04cc04; font-style: italic;"># Writing the top table</span>
  write_tsv(DEGs, paste0(projname, "/", "DEGs.tsv"))
  
  message("+++++ Analysis has completed. +++++")
}




Makevolcano <- <span style="color: #c910c9; font-style: italic;">function</span>(projname,
                        padjlevel = 0.05,
                        uplogFC = 1,
                        downlogFC = -1,
                        ntop = 10) {
  <span style="color: #04cc04; font-style: italic;"># Reading the top table</span>
  table <- read_tsv(paste0(projname, "/", "DEGs.tsv"))
  table <- data.frame(table)
  <span style="color: #04cc04; font-style: italic;"># Colors</span>
  my_pal <- c("#1B9E77","#7570B3", "#E7298A", "#66A61E", "#E6AB02", "#A6761D", "#666666", "#9A7D0A")
  <span style="color: #04cc04; font-style: italic;"># Adding DEG column based on the value of adj.P.Val and logfc</span>
  volcano <- table %>%
    mutate(AdjustedPvalue = -log10(adj.P.Val)) %>%
    mutate(DEGs = "Not") %>%
    mutate(DEGs = ifelse(AdjustedPvalue > -log10(padjlevel) & logFC > uplogFC, "Upregulated", DEGs)) %>%
    mutate(DEGs = ifelse(AdjustedPvalue > -log10(padjlevel) & logFC < downlogFC, "Downregulated", DEGs))

  <span style="color: #04cc04; font-style: italic;"># Creating ggplot object, adding layers and annotation to the plot</span>
  p <- ggplot(data = volcano, aes(x = logFC, y = AdjustedPvalue, color = DEGs, fill = DEGs, text = paste("ID: ", ID,
                                                                                                       "&lt;br&gt;Gene: ", Gene.symbol,
                                                                                                       "&lt;br&gt;Chromosome: ", Chromosome.location,
                                                                                                       "&lt;br&gt;GO function: ", GO.Function))) +
    labs(x= 'log2 (Fold Change)', y = "-log10(Adjusted P-value)") +
    geom_point(size = 1, shape = 21) +
    scale_color_manual(values = c(my_pal)) +
    scale_fill_manual(values = c(paste(my_pal, "66", sep = ""))) +
    theme_classic() +
    theme(axis.text = element_text(family = "Times", size = 15 , colour = "black"),
          axis.text.x = element_text(family = "Times", colour = "black", size = 15),
          axis.text.y = element_text(family = "Times", colour = "black", size = 15),
          plot.subtitle = element_text(family = "Times", size = 20, colour = "black", hjust = 0.5),
          axis.title.y = element_text(family = "Times", size = rel(1.8), angle = 90),
          axis.title.x = element_text(family = "Times", size = rel(1.8), angle = 00)) +
    labs(subtitle = "Volcano plot")
  p <- ggplotly(p, tooltip = "text")
  <span style="color: #04cc04; font-style: italic;"># Calculation the up and down regulated genes number</span>
  up <- volcano[which(volcano$DEGs == "Upregulated"), ]
  up <- up[order(up$logFC, decreasing = T), ]
  down <- volcano[which(volcano$DEGs == "Downregulated"), ]
  down <- down[order(down$logFC, decreasing = F), ]
  <span style="color: #04cc04; font-style: italic;"># Creating the top table (up and down) based on the number specified (ntop)</span>
  top <- rbind(up[1:ntop, ], down[1:ntop, ])
  top <- top[, c(1:6, 11, 7, 8, 9, 10, 12)]
  colnames(top)[7] <- "n.log10(adj.P.Val)"
  <span style="color: #04cc04; font-style: italic;"># Save top table into project folder</span>
  write.csv(top, paste0(projname, "/", "top.csv"), quote = FALSE, row.names = FALSE)
  message(sprintf("Up-regulated genes: %s; Down-regulated genes: %s", nrow(up), nrow(down)))
  <span style="color: #c910c9; font-style: italic;">return</span>(p)
}



DownloadGEO("GSE25055")

#============================ Q1 =========================#
<span style="color: #04cc04; font-style: italic;"># other adjust methods ---> "holm", "hochberg", "hommel", "bonferroni", "BH", "BY", "fdr", "none"</span>

<span style="color: #04cc04; font-style: italic;"># (Lum A, Normal-like subtypes) VS (Basal-like, HER2, Lum B)</span>
DEGanalysis(projname = "GSE25055",
            compare = "subtype",
            groups = list(normal = c("LumA", "Normal"), tumor = c("Basal", "Her2", "LumB")),
            adjust = "fdr")

<span style="color: #04cc04; font-style: italic;"># alpha = 0.01 - logFC 0</span>
volcano1 <- Makevolcano(projname = "GSE25055",
                        padjlevel = 0.01,
                        uplogFC = 1.5,
                        downlogFC = -1.5,
                        ntop = 10)
volcano1

<span style="color: #04cc04; font-style: italic;"># alpha = 0.01 - logFC 1</span>
volcano2 <- Makevolcano(projname = "GSE25055",
                        padjlevel = 0.00001,
                        uplogFC = 1,
                        downlogFC = -1)
volcano2

<span style="color: #04cc04; font-style: italic;"># alpha = 0.05 - logFC 0</span>
volcano3 <- Makevolcano(projname = "GSE25055",
                        padjlevel = 0.05,
                        uplogFC = 0,
                        downlogFC = 0)
volcano3

<span style="color: #04cc04; font-style: italic;"># alpha = 0.05 - logFC 1</span>
volcano4 <- Makevolcano(projname = "GSE25055",
                        padjlevel = 0.05,
                        uplogFC = 1,
                        downlogFC = -1)
volcano4

#============================ Q2 =========================#
<span style="color: #04cc04; font-style: italic;"># Grade 1 vs Grade 3</span>
DEGanalysis(projname = "GSE25055",
            compare = "grade",
            groups = list(normal = c("1"), tumor = c("3")),
            adjust = "fdr")

<span style="color: #04cc04; font-style: italic;"># alpha = 0.01 - logFC 0</span>
volcano5 <- Makevolcano(projname = "GSE25055",
                        padjlevel = 0.01,
                        uplogFC = 0,
                        downlogFC = 0)
volcano5

<span style="color: #04cc04; font-style: italic;"># alpha = 0.01 - logFC 1.5</span>
volcano6 <- Makevolcano(projname = "GSE25055",
                        padjlevel = 0.01,
                        uplogFC = 1.5,
                        downlogFC = -1.5)
volcano6

<span style="color: #04cc04; font-style: italic;"># alpha = 0.05 - logFC 0</span>
volcano7 <- Makevolcano(projname = "GSE25055",
                        padjlevel = 0.05,
                        uplogFC = 0,
                        downlogFC = 0)
volcano7

<span style="color: #04cc04; font-style: italic;"># alpha = 0.01 - logFC 1</span>
volcano8 <- Makevolcano(projname = "GSE25055",
                        padjlevel = 0.01,
                        uplogFC = 1,
                        downlogFC = -1)
volcano8






		</code></pre>
	</div>


	<hr>
	
	<footer class="footer">
	    		<p>&copy; 2023 Mohammad Reza Mohajeri. All rights reserved.</p>
	</footer>

	
	<!-- JavaScript for Drop-down Functionality -->
	<script>
		function toggleDropDown(id) {
			var content = document.getElementById(id);
			content.classList.toggle("active");
		}
	</script>

</body>
</html>
