<!DOCTYPE html>
<html>
<head>
	<meta charset="UTF-8">
	<title>Phase 2 <br> R Code (4-Dataset Analysis - limma)</title>

	
	<!-- 
        This line adds a favicon to the tab of the web browser. 
        The favicon is the small icon displayed next to the page title in the browser tab.
        The 'href' attribute specifies the path to the icon file hosted on GitHub. 
        Ensure the path is correct for the favicon to appear.
	-->
	<link rel="icon" type="image/x-icon" href="https://mohammadrezamohajeri.github.io/Breast-Cancer-Gene-Expression-Biomarker-Analysis/Images/1.ico">


	
	<style>
		/* CSS Styles */
		::selection {
			background-color: yellow; /* Change this to any color you prefer */
			color: black; /* Change the text color if needed */
		}

		
		/* To highlight different parts of the text in red, green, and blue, you can use the <span> element and apply CSS styles to it */ 
		
		.orange-text {
			color: orange;
			font-weight: bold;
		}
		.dark-yellow-text {
			color: hsl(50, 100%, 30%); /* Adjust the hue and saturation values as needed */
			font-weight: bold;
		}
		.brown-text {
			color: hsl(30, 100%, 30%); /* Adjust the hue and saturation values as needed */
			font-weight: bold;
		}
		.green-text {
			color: green;
			font-weight: bold;
		}
		.purple-text {
			color: purple;
			font-weight: bold;
		}
		.blue-text {
			color: blue;
			font-weight: bold;
		}

		/* Header Styles */
		header {
			background-color: #274f4f;
			color: white;
			height: 250px;                 /* Set an explicit height */
			text-align: center;
			position: relative;           /* Add relative positioning here */
			padding: 5px;
		}

		h1 {
			color: white; /* Adjust the font color as desired */
			font-size: 30px; /* Adjust the font size as desired */
			font-weight: bold; /* Add font weight as desired */
			text-align: center; /* Center the header title */
			padding: 0px; /* Add padding around the header title */
			margin-bottom: 20px; /* Add margin below the header title */
		}
		
		hr {
			width: 100%; /* Adjust the width as desired */
		}

		 /* Home Link Style */
	        .home-link {
		        position: absolute;
		        bottom: 10px;
		        left: 10px;
		        text-decoration: none; /* Remove underline */
		        z-index: 1; /* Ensures it appears above other content */
	        }
		
		/* Adjust font size for the code block */
		pre code {
		        font-size: 20px; /* Adjust the font size as desired */
	        }
		
		/* Footer Styles */
		.footer {
		        background-color: #333;
		        color: #fff;
		        padding: 10px;
		        text-align: center;
		        position: fixed;
		        left: 0;
		        bottom: 0;
		        width: 100%;
		}
	</style>
	
</head>
<body>
	<header>
		<!-- the positioning of the "Home" icon -->
	        <a href="https://mohammadrezamohajeri.github.io/Main_Portfolio/index.html" class="home-link">
	            <img src="https://mohammadrezamohajeri.github.io/Main_Portfolio/Pictures/Home_Link_JPG/1.jpg" alt="Home" style="width:50px; height:auto;">
	        </a>
		
		<h1>Phase 2 <br> R Code (4-Dataset Analysis - limma)</h1>
	</header>

	<section>
		<h2>R Code For 4 Datasets (FC/FDR)</h2>
		<pre><code>
<pre><code>
			<!--
<span style="color: #f78102; font-style: italic;">1FUNCTIONFUNCTIONFUNCTIONFUNCTIONFUNCmTIONFUNCTION</span> 
<span style="color: #f78d02; font-style: italic;">2FUNCTIONFUNCTIONFUNCTIONFUNCTIONFUNCTIONFUNCTION</span>
<span style="color: #f79502; font-style: italic;">3FUNCTIONFUNCTIONFUNCTIONFUNCTIONFUNCTIONFUNCTION</span>
<span style="color: #f79502; font-style: italic;">TTTT</span>
<span style="color: #f79d02; font-style: italic;">5FUNCTIONFUNCTIONFUNCTIONFUNCTIONFUNCTIONFUNCTION</span>
<span style="color: #f79d02; font-style: italic;">6FUNCTIONFUNCTIONFUNCTIONFUNCTIONFUNCTIONFUNCTION</span>
<span style="color: #f7aa02; font-style: italic;">7FUNCTIONFUNCTIONFUNCTIONFUNCTIONFUNCTIONFUNCTION</span>
<span style="color: #f7b50c; font-style: italic;">8FUNCTIONFUNCTIONFUNCTIONFUNCTIONFUNCTIONFUNCTION</span>
<span style="color: #f7c707; font-style: italic;">9FUNCTIONFUNCTIONFUNCTIONFUNCTIONFUNCTIONFUNCTION</span>
<span style="color: #e3b707; font-style: italic;">10FUNCTIONFUNCTIONFUNCTIONFUNCTIONFUNCTIONFUNCTION</span>
<span style="color: #c7a108; font-style: italic;">11FUNCTIONFUNCTIONFUNCTIONFUNCTIONFUNCTIONFUNCTION</span>
<span style="color: #ab8b09; font-style: italic;">12FUNCTIONFUNCTIONFUNCTIONFUNCTIONFUNCTIONFUNCTION</span>
<span style="color: #8f750a; font-style: italic;">13FUNCTIONFUNCTIONFUNCTIONFUNCTIONFUNCTIONFUNCTION</span>
<span style="color: #f0e00a; font-style: italic;">14FUNCTIONFUNCTIONFUNCTIONFUNCTIONFUNCTIONFUNCTION</span>
<span style="color: #d4c608; font-style: italic;">15FUNCTIONFUNCTIONFUNCTIONFUNCTIONFUNCTIONFUNCTION</span>
<span style="color: #b8ab04; font-style: italic;">16FUNCTIONFUNCTIONFUNCTIONFUNCTIONFUNCTIONFUNCTION</span>
			-->
			

<span style="color: #999999; font-style: italic;"># Load the GEOquery library for accessing GEO data</span>			
require(GEOquery)
<span style="color: #999999; font-style: italic;"># Load the limma library for differential expression analysis	</span>		
require(limma)
<span style="color: #999999; font-style: italic;"># Load the tidyverse library, which includes several useful packages for data manipulation and visualization</span>		
require(tidyverse)
<span style="color: #999999; font-style: italic;"># Load the plotly library for interactive plots</span>			
require(plotly)
<span style="color: #999999; font-style: italic;"># Load the ggvenn library for creating Venn diagrams using ggplot2</span>			
require(ggvenn)
<span style="color: #999999; font-style: italic;"># Uncomment the line below if devtools package is not installed</span>
<span style="color: #999999; font-style: italic;"># if (!require(devtools)) install.packages("devtools")</span>
<span style="color: #999999; font-style: italic;"># devtools::install_github("yanlinlin82/ggvenn")</span>
require(ggvenn)

<span style="color: #999999; font-style: italic;"># Function to download the specified GEO dataset</span> 
DownloadGEO <- <span style="color: #0548e8; font-style: italic;">function</span>(GSEname = <span style="color: #04cc04; font-style: italic;">"GSE11121"</span>) {
  <span style="color: #999999; font-style: italic;"># Create a folder for the dataset</span>
  dir.create(GSEname, showWarnings = <span style="color: #0548e8; font-style: italic;">FALSE</span>)

  <span style="color: #999999; font-style: italic;"># Increase memory allocation for efficient data processing</span>
  Sys.setenv(<span style="color: #04cc04; font-style: italic;">"VROOM_CONNECTION_SIZE"</span> = <span style="color: #cf2139; font-style: italic;">131072</span> * <span style="color: #cf2139; font-style: italic;">10</span>)
  
  <span style="color: #999999; font-style: italic;"># Download the dataset from the GEO database</span>
  gset <- getGEO(GSEname, GSEMatrix = <span style="color: #0548e8; font-style: italic;">TRUE</span>, AnnotGPL = <span style="color: #0548e8; font-style: italic;">TRUE</span>,
                 destdir = paste0(getwd(), <span style="color: #04cc04; font-style: italic;">"/"</span>, GSEname))
  
  <span style="color: #999999; font-style: italic;"># Check if the dataset was successfully downloaded</span>
  <span style="color: #0548e8; font-style: italic;">if</span> (!<span style="color: #f79502; font-style: italic;">is.null</span>(gset)) {  
    <span style="color: #999999; font-style: italic;"># Extract matrix data from the downloaded dataset</span>
    matrix <- gset[[paste0(GSEname, <span style="color: #04cc04; font-style: italic;">"_series_matrix.txt.gz"</span>)]]@assayData[[<span style="color: #04cc04; font-style: italic;">"exprs"</span>]]
    
    <span style="color: #999999; font-style: italic;"><strong># Perform log2 transformation if the matrix is not already transformed</strong></span>
    <span style="color: #999999; font-style: italic;"># Determine if log2 transformation is required based on the quantiles of the matrix values</span>
    qx <- <span style="color: #f79502; font-style: italic;">as.numeric</span>(quantile(matrix, <span style="color: #f79502; font-style: italic;">c</span>(<span style="color: #cf2139; font-style: italic;">0.</span>, <span style="color: #cf2139; font-style: italic;">0.25</span>, <span style="color: #cf2139; font-style: italic;">0.5</span>, <span style="color: #cf2139; font-style: italic;">0.75</span>, <span style="color: #cf2139; font-style: italic;">0.99</span>, <span style="color: #cf2139; font-style: italic;">1.0</span>), na.rm = <span style="color: #0548e8; font-style: italic;">TRUE</span>))
    LogC <- (qx[<span style="color: #cf2139; font-style: italic;">5</span>] > <span style="color: #cf2139; font-style: italic;">100</span>) || (qx[<span style="color: #cf2139; font-style: italic;">6</span>] - qx[<span style="color: #cf2139; font-style: italic;">1</span>] > <span style="color: #cf2139; font-style: italic;">50</span> && qx[<span style="color: #cf2139; font-style: italic;">2</span>] > <span style="color: #cf2139; font-style: italic;">0</span>)
    
    <span style="color: #999999; font-style: italic;"># Check if log2 transformation is required</span>
    <span style="color: #0548e8; font-style: italic;">if</span> (LogC) {
      <span style="color: #999999; font-style: italic;"># Replace non-positive values with NA</span>
      matrix[matrix <= <span style="color: #cf2139; font-style: italic;">0</span>] <- <span style="color: #0548e8; font-style: italic;">NA</span>
      <span style="color: #999999; font-style: italic;"># Perform log2 transformation on the matrix</span>
      matrix <- log2(matrix)
    }
    
    <span style="color: #999999; font-style: italic;"><strong># Add row names to the matrix:</strong></span>
    <span style="color: #999999; font-style: italic;"># Add row names as the first column of the matrix using cbind()</span>
    matrix <- cbind(rownames(matrix), matrix)
    <span style="color: #999999; font-style: italic;"># Set the column name of the first column as "ID" using colnames()</span>
    colnames(matrix)[<span style="color: #cf2139; font-style: italic;">1</span>] <- <span style="color: #04cc04; font-style: italic;">"ID"</span>
    
    <span style="color: #999999; font-style: italic;"><strong># Extract phenotype information from the dataset:</strong></span>
    <span style="color: #999999; font-style: italic;"># Extract the phenotype information from the dataset</span>
    phenotype <- gset[[paste0(GSEname, <span style="color: #04cc04; font-style: italic;">"_series_matrix.txt.gz"</span>)]]@phenoData@data  
    <span style="color: #999999; font-style: italic;"># Add row names as the first column of the phenotype data using cbind()</span>
    phenotype <- cbind(rownames(phenotype), phenotype)  
    <span style="color: #999999; font-style: italic;"># Set the column name of the first column as "ID" using colnames()</span>
    colnames(phenotype)[<span style="color: #cf2139; font-style: italic;">1</span>] <- <span style="color: #04cc04; font-style: italic;">"ID" </span> 
    
    <span style="color: #999999; font-style: italic;"># Extract annotation information from the dataset</span>
    annot <- gset[[paste0(GSEname, <span style="color: #04cc04; font-style: italic;">"_series_matrix.txt.gz"</span>)]]@featureData@data
    
    <span style="color: #999999; font-style: italic;"><strong># Write the extracted variables to files:</strong></span>
    <span style="color: #999999; font-style: italic;"># Write the 'matrix' variable to a CSV file</span>
    write.csv(matrix, paste0(GSEname, <span style="color: #04cc04; font-style: italic;">"/"</span>, <span style="color: #04cc04; font-style: italic;">"matrix.csv"</span>), <span style="color: #f79502; font-style: italic;">quote</span> = <span style="color: #0548e8; font-style: italic;">FALSE</span>, row.names = <span style="color: #0548e8; font-style: italic;">FALSE</span>)
    <span style="color: #999999; font-style: italic;"># Write the 'phenotype' variable to a TSV file</span>
    write_tsv(phenotype, paste0(GSEname, <span style="color: #04cc04; font-style: italic;">"/"</span>, <span style="color: #04cc04; font-style: italic;">"phenotype.tsv"</span>))
    <span style="color: #999999; font-style: italic;"># Write the 'annot' variable to a TSV file</span>
    write_tsv(annot, paste0(GSEname, <span style="color: #04cc04; font-style: italic;">"/"</span>, <span style="color: #04cc04; font-style: italic;">"annot.tsv"</span>))
    
    <span style="color: #999999; font-style: italic;"># Remove unnecessary variables to free up memory</span>
    rm(gset, matrix, phenotype, annot)

    <span style="color: #999999; font-style: italic;"># Display a completion message</span>
    message(<span style="color: #04cc04; font-style: italic;">"+++++ Dataset is ready to be analyzed. +++++"</span>)
  }
}


<span style="color: #999999; font-style: italic;"># Function to perform differential expression analysis on the specified GEO dataset.</span>
DEGanalysis <- <span style="color: #0548e8; font-style: italic;">function</span>(projname,
                        compare = <span style="color: #04cc04; font-style: italic;">"subtype"</span>,
                        groups,
                        adjust = <span style="color: #04cc04; font-style: italic;">"fdr"</span>) {
  <span style="color: #999999; font-style: italic;"><strong># Reading the files:</strong></span>
  <span style="color: #999999; font-style: italic;"># Read the matrix file and store it in a variable</span>
  matrix <- read_csv(paste0(projname, <span style="color: #04cc04; font-style: italic;">"/"</span>, <span style="color: #04cc04; font-style: italic;">"matrix.csv"</span>), )
  <span style="color: #999999; font-style: italic;"># Convert the matrix into a data frame</span>
  matrix <- data.frame(matrix)
  <span style="color: #999999; font-style: italic;"># Set row names of the matrix</span>
  rownames(matrix) <- matrix[, <span style="color: #cf2139; font-style: italic;">1</span>]
  <span style="color: #999999; font-style: italic;"># Remove the first column of the matrix</span>
  matrix <- matrix[, <span style="color: #cf2139; font-style: italic;">-1</span>]
  
  <span style="color: #999999; font-style: italic;"># Read the phenotype file and store it in a variable</span>
  phenotype <- read_tsv(paste0(projname, <span style="color: #04cc04; font-style: italic;">"/"</span>, <span style="color: #04cc04; font-style: italic;">"phenotype.tsv"</span>))
  <span style="color: #999999; font-style: italic;"># Convert the phenotype into a data frame</span>
  phenotype <- data.frame(phenotype)
  <span style="color: #999999; font-style: italic;"># Set row names of the phenotype</span>
  rownames(phenotype) <- phenotype[, <span style="color: #cf2139; font-style: italic;"1></span>]
  <span style="color: #999999; font-style: italic;"># Remove the first column of the phenotype</span>
  phenotype <- phenotype[, <span style="color: #cf2139; font-style: italic;">-1</span>]
  
  <span style="color: #999999; font-style: italic;"># Read the annotation file and store it in a variable</span>
  annot <- read_tsv(paste0(projname, <span style="color: #04cc04; font-style: italic;">"/"</span>, <span style="color: #04cc04; font-style: italic;">"annot.tsv"</span>))
  <span style="color: #999999; font-style: italic;"># Convert the annotation into a data frame</span>
  annot <- data.frame(annot)
  <span style="color: #999999; font-style: italic;"># Set row names of the annotation</span>
  rownames(annot) <- annot[, <span style="color: #cf2139; font-style: italic;">1</span>]
  <span style="color: #999999; font-style: italic;"># Remove the first column of the annotation</span>
  annot <- annot[, <span style="color: #cf2139; font-style: italic;">-1</span>]
  
  <span style="color: #999999; font-style: italic;"># Extract sample IDs based on user-specified groups for differential analysis</span>
  <span style="color: #0548e8; font-style: italic;">if</span> (compare == <span style="color: #04cc04; font-style: italic;">"subtype"</span>) {
    
    <span style="color: #999999; font-style: italic;"># Preparing the breast cancer subtype</span>
    phecoln <- grep(<span style="color: #04cc04; font-style: italic;">"pam50"</span>, phenotype)
    <span style="color: #0548e8; font-style: italic;">if</span> (<span style="color: #f79502; font-style: italic;">length</span>(phecoln) > <span style="color: #cf2139; font-style: italic;">0</span>) {
      <span style="color: #999999; font-style: italic;"># Prepare breast cancer subtypes into a dataframe</span>
      pheno <- data.frame(phenotype[, phecoln])
      <span style="color: #999999; font-style: italic;"># Extract relevant columns from the phenotype file for breast cancer subtypes and rename them</span>
      colnames(pheno) <- <span style="color: #04cc04; font-style: italic;">"subtype"</span>
      pheno$subtype <- gsub(<span style="color: #04cc04; font-style: italic;">"pam50.+:"</span>, <span style="color: #04cc04; font-style: italic;">""</span>, pheno$subtype)
      rownames(pheno) <- rownames(phenotype)
      <span style="color: #999999; font-style: italic;"># Store the extracted subtypes in a list, grouped by user-specified group names</span>
      grouplis <- <span style="color: #f79502; font-style: italic;">list</span>()
      <span style="color: #999999; font-style: italic;"># Iterate over each group specified by the user</span>
      <span style="color: #0548e8; font-style: italic;">for</span> (i <span style="color: #0548e8; font-style: italic;">in</span> <span style="color: #f79502; font-style: italic;">seq_along</span>(groups)) {
        <span style="color: #999999; font-style: italic;"># Create a regular expression pattern by concatenating group elements with a "|"</span>
        name <- paste0(groups[[i]], collapse = <span style="color: #04cc04; font-style: italic;">"|"</span>)
        <span style="color: #999999; font-style: italic;"># Find the row numbers in the "subtype" column of the phenotype data that match the pattern</span>
        rownum <- grep(name, pheno$subtype)
        <span style="color: #999999; font-style: italic;"># Store the row names corresponding to the matched rows in the grouplis list</span>
        grouplis[[<span style="color: #f79502; font-style: italic;">names</span>(groups)[i]]] <- rownames(pheno)[rownum]
      }
    }
  } <span style="color: #0548e8; font-style: italic;">else</span> {
    
    <span style="color: #999999; font-style: italic;"><strong># Preparing the breast cancer grades:</strong></span>
    <span style="color: #999999; font-style: italic;"># Find columns containing "grade:" in the 'phenotype' data frame</span>
    gradecoln <- grep(<span style="color: #04cc04; font-style: italic;">"grade:"</span>, phenotype)
    
    <span style="color: #999999; font-style: italic;"># Check if there are any columns with "grade:" found</span>
    <span style="color: #0548e8; font-style: italic;">if</span> (<span style="color: #f79502; font-style: italic;">length</span>(gradecoln) > <span style="color: #cf2139; font-style: italic;">0</span>) {
      <span style="color: #999999; font-style: italic;"># Prepare breast cancer grades into a dataframe</span>
      grade <- data.frame(phenotype[, gradecoln])
      
      <span style="color: #999999; font-style: italic;"><strong># Extract relevant columns from phenotype file for breast cancer grades and rename them:</strong></span>
      <span style="color: #999999; font-style: italic;"># Rename the column in 'grade' data frame to "grade"</span>
      colnames(grade) <- <span style="color: #04cc04; font-style: italic;">"grade"</span>
      <span style="color: #999999; font-style: italic;"># Remove unwanted text patterns from the 'grade' column using regular expressions</span>
      grade$grade <- gsub(<span style="color: #04cc04; font-style: italic;">"grade:|grade: |B-R grade: "</span>, <span style="color: #04cc04; font-style: italic;">""</span>, grade$grade)
      grade$grade <- gsub(<span style="color: #04cc04; font-style: italic;">"=.+"</span>, <span style="color: #04cc04; font-style: italic;">""</span>, grade$grade)
      
      <span style="color: #999999; font-style: italic;"># Identify rows containing the pattern "III" in the 'grade' column</span>
      greek <- grep(<span style="color: #04cc04; font-style: italic;">"III"</span>, grade$grade)
      <span style="color: #0548e8; font-style: italic;">if</span> (<span style="color: #f79502; font-style: italic;">length</span>(greek) > <span style="color: #cf2139; font-style: italic;">0</span>) {
        
        <span style="color: #999999; font-style: italic;"><strong># Convert grade names to numeric values:</strong></span>
        <span style="color: #999999; font-style: italic;"># Identify rows with grade "I"</span>
        one <- which(factor(grade$grade) == <span style="color: #04cc04; font-style: italic;">"I"</span>)
        <span style="color: #999999; font-style: italic;"># Identify rows with grade "II"</span>
        two <- which(factor(grade$grade) == <span style="color: #04cc04; font-style: italic;">"II"</span>)
        <span style="color: #999999; font-style: italic;"># Identify rows with grade "III"</span>
        three <- which(factor(grade$grade) == <span style="color: #04cc04; font-style: italic;">"III"</span>)
        
        <span style="color: #999999; font-style: italic;"># Convert grade values to numeric: "I" -> 1, "II" -> 2, "III" -> 3</span>
        grade[one, ] <- <span style="color: #04cc04; font-style: italic;">"1"</span>
        grade[two, ] <- <span style="color: #04cc04; font-style: italic;">"2"</span>
        grade[three, ] <- <span style="color: #04cc04; font-style: italic;">"3"</span>
      }
      
      <span style="color: #999999; font-style: italic;"># Set row names of 'grade' data frame to row names of 'phenotype'</span>
      rownames(grade) <- rownames(phenotype)
      
      <span style="color: #999999; font-style: italic;"># Store the extracted grades in a list, grouped by user-specified group names</span>
      grouplis <- <span style="color: #f79502; font-style: italic;">list</span>()
      <span style="color: #0548e8; font-style: italic;">for</span> (i <span style="color: #0548e8; font-style: italic;">in</span> <span style="color: #f79502; font-style: italic;">seq_along</span>(groups)) {
        <span style="color: #999999; font-style: italic;"># Generate a group name by collapsing the elements of 'groups' with "|"</span>
        name <- paste0(groups[[i]], collapse = <span style="color: #04cc04; font-style: italic;">"|"</span>)
        <span style="color: #999999; font-style: italic;"># Find the row numbers in 'grade' where the group name is present in the 'grade' column</span>
        rownum <- grep(name, grade$grade)
        <span style="color: #999999; font-style: italic;"># Store the corresponding row names in the 'grouplis' list under the appropriate group name</span>
        grouplis[[<span style="color: #f79502; font-style: italic;">names</span>(groups)[i]]] <- rownames(grade)[rownum]
      }
    }
  }
  
  <span style="color: #999999; font-style: italic;"><strong># Rename the matrix file according to the groups name specified by the user:</strong></span>
  <span style="color: #999999; font-style: italic;"># Iterate over each group in grouplis</span>		
  <span style="color: #0548e8; font-style: italic;">for</span> (i <span style="color: #0548e8; font-style: italic;">in</span> <span style="color: #f79502; font-style: italic;">seq_along</span>(grouplis)) {
    <span style="color: #999999; font-style: italic;"># Generate a group name by collapsing the elements of 'grouplis[[i]]' with "|"</span>
    matname <- paste0(grouplis[[i]], collapse = <span style="color: #04cc04; font-style: italic;">"|"</span>)
    <span style="color: #999999; font-style: italic;"># Find the column numbers in 'matrix' where the group name is present in the column names</span>
    matcoln <- grep(matname, colnames(matrix))
    <span style="color: #999999; font-style: italic;"># Rename the columns of 'matrix' with the group name followed by a numeric index</span>
    colnames(matrix)[matcoln] <- paste0(<span style="color: #f79502; font-style: italic;">names</span>(grouplis)[i], <span style="color: #cf2139; font-style: italic;">1</span>:<span style="color: #f79502; font-style: italic;">length</span>(grouplis[[i]]))
  }
   
  <span style="color: #999999; font-style: italic;"><strong># Keep only the renamed samples in the matrix and order them:</strong></span>
  <span style="color: #999999; font-style: italic;"># Subset and reorder the 'matrix' data frame based on the group names</span>
  matnames <- grep(paste0(<span style="color: #f79502; font-style: italic;">names</span>(grouplis), collapse = <span style="color: #04cc04; font-style: italic;">"|"</span>), colnames(matrix))
  <span style="color: #999999; font-style: italic;"># Subset the 'matrix' data frame to include only the columns corresponding to the renamed samples</span>
  matrix <- matrix[, matnames]
  <span style="color: #999999; font-style: italic;"># Reorder the columns of the 'matrix' data frame based on the column names</span>
  matrix <- matrix[, order(colnames(matrix))]
  
  <span style="color: #999999; font-style: italic;"><strong># Creating a factor of 0 and 1 according to the length of groups specified by the user:</strong></span>
  <span style="color: #999999; font-style: italic;"># Create an empty list to store the factors</span>
  gname <- <span style="color: #f79502; font-style: italic;">list</span>()
  <span style="color: #999999; font-style: italic;"># Iterate over each group in grouplis</span>
  <span style="color: #0548e8; font-style: italic;">for</span> (i <span style="color: #0548e8; font-style: italic;">in</span> <span style="color: #f79502; font-style: italic;">seq_along</span>(grouplis)) {
    <span style="color: #999999; font-style: italic;"># Count the number of columns in matrix corresponding to the group</span>
    number <- <span style="color: #f79502; font-style: italic;">length</span>(grep(<span style="color: #f79502; font-style: italic;">names</span>(grouplis)[i], colnames(matrix)))
    <span style="color: #999999; font-style: italic;"># Create a factor variable with values i - 1 repeated number times</span>
    gname[[i]] <- factor(<span style="color: #f79502; font-style: italic;">rep</span>(i - <span style="color: #cf2139; font-style: italic;">1</span>, number))
  }
  <span style="color: #999999; font-style: italic;"># Convert the list of factors to a single vector</span>
  grouping <- unlist(gname)
  
  <span style="color: #999999; font-style: italic;"><strong># Create a design matrix for the linear model:</strong></span>
  <span style="color: #999999; font-style: italic;"># Create a design matrix with the factor variables as predictors</span>
  design <- model.matrix(~ i - <span style="color: #cf2139; font-style: italic;">0</span> + grouping)
  <span style="color: #999999; font-style: italic;"># Assign column names to the design matrix based on the group names</span>
  colnames(design) <-<span style="color: #f79502; font-style: italic;">names</span>(grouplis)
  
  <span style="color: #999999; font-style: italic;"><strong># Fit the linear model to the data:</strong></span>
  <span style="color: #999999; font-style: italic;"># Fit the linear model using the matrix and design matrix</span>
  fit1 <- lmFit(matrix, design)
  <span style="color: #999999; font-style: italic;"># Get the number of group names</span>
  l <- <span style="color: #f79502; font-style: italic;">length</span>(<span style="color: #f79502; font-style: italic;">names</span>(grouplis))
  <span style="color: #999999; font-style: italic;"># Create a string representation of group names for contrasts</span>
  cts <- paste0(<span style="color: #f79502; font-style: italic;">names</span>(grouplis)[l:<span style="color: #cf2139; font-style: italic;">1</span>], collapse = <span style="color: #04cc04; font-style: italic;">"-"</span>)
  <span style="color: #999999; font-style: italic;"># Create contrasts for pairwise comparisons of groups</span>
  cont.matrix <- makeContrasts(contrasts = cts, levels = design)
  <span style="color: #999999; font-style: italic;"># Fit the contrasts to the model</span>
  fit2 <- contrasts.fit(fit1, cont.matrix)
  <span style="color: #999999; font-style: italic;"># Perform empirical Bayes moderation on the fitted model</span>
  fit2 <- eBayes(fit2)
  
  <span style="color: #999999; font-style: italic;"># Generate a top table of differentially expressed genes</span>
  DEGs <- topTable(fit2, adjust = adjust, number = <span style="color: #0548e8; font-style: italic;">Inf</span>)
  <span style="color: #999999; font-style: italic;"># Add row names as a separate column in the DEGs table</span>
  DEGs <- cbind(rownames(DEGs), DEGs)
  <span style="color: #999999; font-style: italic;"># Rename the first column as "ID"</span>
  colnames(DEGs)[<span style="color: #cf2139; font-style: italic;">1</span>] <- <- <span style="color: #04cc04; font-style: italic;">"ID"</span>
  
  <span style="color: #999999; font-style: italic;"><strong># Preparing the annotation variable:</strong></span>
  <span style="color: #999999; font-style: italic;"># Extract relevant columns from the 'annot' data frame</span>
  annotation <- annot[, <span style="color: #f79502; font-style: italic;">c</span>(<span style="color: #04cc04; font-style: italic;">"Gene.symbol"</span>, <span style="color: #04cc04; font-style: italic;">"Chromosome.location"</span>, <span style="color: #04cc04; font-style: italic;">"GO.Function"</span>)]
  <span style="color: #999999; font-style: italic;"># Add row names as a separate column in the 'annotation' data frame</span>
  annotation <- cbind(rownames(annotation), annotation)
  <span style="color: #999999; font-style: italic;"># Rename the first column as "ID"</span>
  colnames(annotation)[<span style="color: #cf2139; font-style: italic;">1</span>] <- <- <span style="color: #04cc04; font-style: italic;">"ID"</span>
  
  <span style="color: #999999; font-style: italic;"><strong># Making sure that both DEGs and annotation row names are the same:</strong></span>
  <span style="color: #999999; font-style: italic;"># Sort the 'annotation' data frame by the "ID" column</span>
  annotation <- annotation[order(annotation$ID), ]
  <span style="color: #999999; font-style: italic;"># Sort the 'DEGs' data frame by the "ID" column</span>
  DEGs <- DEGs[order(DEGs$ID), ]
  
  <span style="color: #999999; font-style: italic;"># Merge the top table with the annotation data </span>
  DEGs <- cbind(DEGs, annotation[, <span style="color: #f79502; font-style: italic;">c</span>(<span style="color: #cf2139; font-style: italic;">2</span>, <span style="color: #cf2139; font-style: italic;">3</span>, <span style="color: #cf2139; font-style: italic;">4</span>)])
  
  <span style="color: #999999; font-style: italic;"><strong># Removing missing gene symbols:</strong></span>
  <span style="color: #999999; font-style: italic;"># Find the indices of rows with missing gene symbols</span>
  genemissing <- which(<span style="color: #f79502; font-style: italic;">is.na</span>(DEGs$Gene.symbol) == <span style="color: #0548e8; font-style: italic;">TRUE</span>)
  <span style="color: #999999; font-style: italic;"># Remove the rows with missing gene symbols from the 'DEGs' data frame</span>
  DEGs <- DEGs[-genemissing, ]
  
  <span style="color: #999999; font-style: italic;"># Write the resulting top table to a file</span>
  write_tsv(DEGs, paste0(projname, <span style="color: #04cc04; font-style: italic;">"/"</span>, <span style="color: #04cc04; font-style: italic;">"DEGs.tsv"</span>))
  
  <span style="color: #999999; font-style: italic;"># Display a completion message</span>
  message(<span style="color: #04cc04; font-style: italic;">"+++++ Analysis has completed. +++++"</span>)
}



<span style="color: #999999; font-style: italic;"># Function to generates volcano plots for the specified GEO datasets.</span>
Makevolcano <- <span style="color: #0548e8; font-style: italic;">function</span>(projname,
                        padjlevel = <span style="color: #cf2139; font-style: italic;">0.05</span>,
                        uplogFC = <span style="color: #cf2139; font-style: italic;">1</span>,
                        downlogFC = <span style="color: #cf2139; font-style: italic;">-1</span>,
                        ntop = <span style="color: #cf2139; font-style: italic;">10</span>,
                        voltype = <span style="color: #04cc04; font-style: italic;">"plotly"</span>) {
  <span style="color: #999999; font-style: italic;"># Read the tab-separated DEGs file into a data frame</span>
  table <- read_tsv(paste0(projname, <span style="color: #04cc04; font-style: italic;">"/"</span>, <span style="color: #04cc04; font-style: italic;">"DEGs.tsv"</span>))
  table <- data.frame(table)
  
  <span style="color: #999999; font-style: italic;"># Define a vector of color codes</span>
  my_pal <- <span style="color: #f79502; font-style: italic;">c</span>(<span style="color: #1B9E77; font-style: italic;">"#1B9E77"</span>, <span style="color: #7570B3; font-style: italic;">"#7570B3"</span>, <span style="color: #E7298A; font-style: italic;">"#E7298A"</span>, <span style="color: #66A61E; font-style: italic;">"#66A61E"</span>, <span style="color: #E6AB02; font-style: italic;">"#E6AB02"</span>, <span style="color: #A6761D; font-style: italic;">"#A6761D"</span>, <span style="color: #666666; font-style: italic;">"#666666"</span>, <span style="color: #9A7D0A; font-style: italic;">"#9A7D0A"</span>)
  
  <span style="color: #999999; font-style: italic;"><strong># Add DEG column based on the value of adj.P.Val and logFC:</strong></span>
  volcano <- table %>%
    <span style="color: #999999; font-style: italic;"># Calculate AdjustedPvalue as the negative logarithm (base 10) of adj.P.Val</span>
    mutate(AdjustedPvalue = -log10(adj.P.Val)) %>%
    <span style="color: #999999; font-style: italic;"># Initialize DEG column with "Non-Significant" for all rows</span>
    mutate(DEG = <span style="color: #04cc04; font-style: italic;">"NotSignificant"</span>) %>%
    <span style="color: #999999; font-style: italic;"># Update DEG based on conditions for upregulation</span>
    mutate(DEG = ifelse(AdjustedPvalue > -log10(padjlevel) & logFC > uplogFC, "Upregulated", DEG)) %>%
    <span style="color: #999999; font-style: italic;"># Update DEG based on conditions for downregulation</span>
    mutate(DEG = ifelse(AdjustedPvalue > -log10(padjlevel) & logFC < downlogFC, "Downregulated", DEG))
  
  <span style="color: #999999; font-style: italic;"><strong># Separate the DEGs into three categories: Upregulated, Downregulated, Non-Significant:</strong></span>
  <span style="color: #999999; font-style: italic;"># Subset the 'volcano' data frame to select only the rows where the 'DEG' column is equal to "Upregulated".</span>
  volcano_up <- volcano[which(volcano$DEG == <span style="color: #04cc04; font-style: italic;">"Upregulated"</span>), ]
  <span style="color: #999999; font-style: italic;"># Sort the 'volcano_up' data frame in descending order based on the 'logFC' column.</span>
  volcano_up <- volcano_up[order(volcano_up$logFC, decreasing = <span style="color: #f79502; font-style: italic;">T</span>), ]
  <span style="color: #999999; font-style: italic;"># Subset the 'volcano' data frame to select only the rows where the 'DEG' column is equal to "Downregulated".</span>
  volcano_down <- volcano[which(volcano$DEG == <span style="color: #04cc04; font-style: italic;">"Downregulated"</span>), ]
  <span style="color: #999999; font-style: italic;"># Sort the 'volcano_down' data frame in ascending order based on the 'logFC' column.</span>
  volcano_down <- volcano_down[order(volcano_down$logFC, decreasing = <span style="color: #f79502; font-style: italic;">F</span>), ]
  <span style="color: #999999; font-style: italic;"># Subset the 'volcano' data frame to select only the rows where the 'DEG' column is equal to "Non-Significant".</span>
  volcano_NA <- volcano[which(volcano$DEG == <span style="color: #04cc04; font-style: italic;">"NotSignificant"</span>), ]
  
  <span style="color: #999999; font-style: italic;"><strong># Combine the DEGs into a single data frame:</strong></span>
  <span style="color: #999999; font-style: italic;"># Combine the 'volcano_up', 'volcano_down', and 'volcano_NA' data frames vertically using the 'rbind' function.</span>
  volcano_all <- rbind(volcano_up, volcano_down, volcano_NA)
  
  <span style="color: #999999; font-style: italic;"><strong># Replace the top probes with the word "Top" in the DEGs column:</strong></span>
  <span style="color: #999999; font-style: italic;"># Create a vector 'ntop_rep' with length 'ntop' where each element is "Top".</span>
  ntop_rep <- <span style="color: #f79502; font-style: italic;">rep</span>(<span style="color: #04cc04; font-style: italic;">"Top"</span>, ntop)
  <span style="color: #999999; font-style: italic;"># Create a new column 'DEGs' in the 'volcano_all' data frame.</span>
  volcano_all$DEGs <- <span style="color: #f79502; font-style: italic;">c</span>(ntop_rep, volcano_up$DEG[-<span style="color: #f79502; font-style: italic;">c</span>(<span style="color: #cf2139; font-style: italic;">1</span>:ntop)], ntop_rep, volcano_down$DEG[-<span style="color: #f79502; font-style: italic;">c</span>(<span style="color: #cf2139; font-style: italic;">1</span>:ntop)], volcano_NA$DEG)
  
  <span style="color: #999999; font-style: italic;"><strong># Label only top genes:</strong></span>
  <span style="color: #999999; font-style: italic;"># Assign the gene symbol as a label for top genes</span>
  volcano_all$Top <- ifelse(volcano_all$DEGs == <span style="color: #04cc04; font-style: italic;">"Top"</span>, volcano_all$Gene.symbol, <span style="color: #04cc04; font-style: italic;">""</span>)
  
  <span style="color: #999999; font-style: italic;"># Plotting either with plotly or ggplot</span>
  <span style="color: #0548e8; font-style: italic;">if</span> (voltype == <span style="color: #04cc04; font-style: italic;">"plotly"</span>) {
    <span style="color: #999999; font-style: italic;"># Create a ggplot object with layers and annotations</span>
    p <- ggplot(data = volcano_all, aes(x = logFC, y = AdjustedPvalue, color = DEGs, fill = DEGs, text = paste("ID: ", ID,
													       <span style="color: #999999; font-style: italic;"># Tooltip text for gene symbol</span>
	                                                                                                       <span style="color: #04cc04; font-style: italic;">"&lt;br&gt; Gene: "</span>, Gene.symbol,
	                                                                                                       <span style="color: #999999; font-style: italic;"># Tooltip text for chromosome location</span>
	                                                                                                       <span style="color: #04cc04; font-style: italic;">"&lt;br&gt; Chromosome: "</span>, Chromosome.location,
	                                                                                                       <span style="color: #999999; font-style: italic;"># Tooltip text for GO function</span>
	                                                                                                       <span style="color: #04cc04; font-style: italic;">"&lt;br&gt; GO function: "</span>, GO.Function))) +
      <span style="color: #999999; font-style: italic;"># Set the x-axis label, y-axis label, and plot title</span>
      labs(x = <span style="color: #04cc04; font-style: italic;">'log2 (Fold Change)'</span>, y = <span style="color: #04cc04; font-style: italic;">"-log10(Adjusted P-value)"</span>, title = paste0(<span style="color: #04cc04; font-style: italic;">"Volcano plot-"</span>, projname)) +
      <span style="color: #999999; font-style: italic;"># Add points to the plot with specified size and shape</span>
      geom_point(size = <span style="color: #cf2139; font-style: italic;">1</span>, shape = <span style="color: #cf2139; font-style: italic;">21</span>) +
      <span style="color: #999999; font-style: italic;"># Set the color scale of the plot manually</span>
      scale_color_manual(values = <span style="color: #f79502; font-style: italic;">c</span>(my_pal)) +
      <span style="color: #999999; font-style: italic;"># Set the fill color scale of the plot manually</span>
      scale_fill_manual(values = <span style="color: #f79502; font-style: italic;">c</span>(paste(my_pal, <span style="color: #04cc04; font-style: italic;">"66"</span>, sep = <span style="color: #04cc04; font-style: italic;">""</span>))) +
      <span style="color: #999999; font-style: italic;"># Apply the "classic" theme to the plot</span>
      theme_classic() +
      <span style="color: #999999; font-style: italic;"># Set axis text font and size</span>
      theme(axis.text = element_text(family = <span style="color: #04cc04; font-style: italic;">"Arial"</span>, size = <span style="color: #cf2139; font-style: italic;">24</span>, colour = <span style="color: #04cc04; font-style: italic;">"black"</span>),
            <span style="color: #999999; font-style: italic;"># Set x-axis text color and size</span>
            axis.text.x = element_text(family = <span style="color: #04cc04; font-style: italic;">"Arial"</span>, colour = <span style="color: #04cc04; font-style: italic;">"black"</span>, size = <span style="color: #cf2139; font-style: italic;">24</span>),
            <span style="color: #999999; font-style: italic;"># Set y-axis text color and size</span>
            axis.text.y = element_text(family = <span style="color: #04cc04; font-style: italic;">"Arial"</span>, colour = <span style="color: #04cc04; font-style: italic;">"black"</span>, size = <span style="color: #cf2139; font-style: italic;">24</span>),
            <span style="color: #999999; font-style: italic;"># Set subtitle font, size, and color</span>
            plot.subtitle = element_text(family = <span style="color: #04cc04; font-style: italic;">"Arial"</span>, size = <span style="color: #cf2139; font-style: italic;">24</span>, colour = <span style="color: #04cc04; font-style: italic;">"black"</span>, hjust = <span style="color: #cf2139; font-style: italic;">0.5</span>),
            <span style="color: #999999; font-style: italic;"># Set y-axis title font, size, and angle</span>
            axis.title.y = element_text(family = <span style="color: #04cc04; font-style: italic;">"Arial"</span>, size = <span style="color: #cf2139; font-style: italic;">24</span>, angle = <span style="color: #cf2139; font-style: italic;">90</span>),
            <span style="color: #999999; font-style: italic;"># Set x-axis title font, size, and angle</span>
            axis.title.x = element_text(family = <span style="color: #04cc04; font-style: italic;">"Arial"</span>, size = <span style="color: #cf2139; font-style: italic;">24</span>, angle = <span style="color: #cf2139; font-style: italic;">00</span>))
    
    <span style="color: #999999; font-style: italic;"># Convert ggplot object to plotly object and customize layout</span>
    p <- ggplotly(p, tooltip = <span style="color: #04cc04; font-style: italic;">"text"</span>) %>%
      <span style="color: #999999; font-style: italic;"># Set the title of the plot</span>
      layout(title = paste0(<span style="color: #04cc04; font-style: italic;">"Volcano plot-"</span>, projname),
             <span style="color: #999999; font-style: italic;"># Customize font properties</span>
             font = <span style="color: #f79502; font-style: italic;">list</span>(family = <span style="color: #04cc04; font-style: italic;">"Arial"</span>, color = <span style="color: #04cc04; font-style: italic;">"black"</span>, size = <span style="color: #cf2139; font-style: italic;">24</span>),
             <span style="color: #999999; font-style: italic;"># Customize legend properties</span>
             legend = <span style="color: #f79502; font-style: italic;">list</span>(family = <span style="color: #04cc04; font-style: italic;">"Arial"</span>, color = <span style="color: #04cc04; font-style: italic;">"black"</span>, size = <span style="color: #cf2139; font-style: italic;">20</span>, itemclick = <span style="color: #04cc04; font-style: italic;">"toggle"</span>))
  } <span style="color: #0548e8; font-style: italic;">else</span> {
    <span style="color: #999999; font-style: italic;"># Create a ggplot object with layers and annotations:</span>
    p <- ggplot(data = volcano_all, aes(x = logFC, y = AdjustedPvalue, color = DEGs, fill = DEGs, label = Top)) + 
      <span style="color: #999999; font-style: italic;"># Set the labels for x and y axes</span>
      labs(x = <span style="color: #04cc04; font-style: italic;">'log2 (Fold Change)'</span>, y = <span style="color: #04cc04; font-style: italic;">"-log10(Adjusted P-value)"</span>) +
      <span style="color: #999999; font-style: italic;"># Add points to the plot</span>
      geom_point(size = <span style="color: #cf2139; font-style: italic;">1</span>, shape = <span style="color: #cf2139; font-style: italic;">21</span>) +
      <span style="color: #999999; font-style: italic;"># Add text annotations to the plot</span>
      geom_text(check_overlap = <span style="color: #f79502; font-style: italic;">F</span>, vjust = <span style="color: #cf2139; font-style: italic;">0.1</span>, nudge_y = <span style="color: #cf2139; font-style: italic;">0.4</span>) +
      <span style="color: #999999; font-style: italic;"># Set manual color scale for DEGs</span>
      scale_color_manual(values = <span style="color: #f79502; font-style: italic;">c</span>(my_pal)) +
      <span style="color: #999999; font-style: italic;"># Set manual fill scale for DEGs</span>
      scale_fill_manual(values = <span style="color: #f79502; font-style: italic;">c</span>(paste(my_pal, <span style="color: #04cc04; font-style: italic;">"66"</span>, sep = <span style="color: #04cc04; font-style: italic;">""</span>))) +
      <span style="color: #999999; font-style: italic;"># Use classic theme for the plot</span>
      theme_classic() +
     
      theme(
        <span style="color: #999999; font-style: italic;"># Customize axis text properties</span>
        axis.text = element_text(family = <span style="color: #04cc04; font-style: italic;">"Arial"</span>, size = <span style="color: #cf2139; font-style: italic;">24</span>, colour = <span style="color: #04cc04; font-style: italic;">"black"</span>),
        <span style="color: #999999; font-style: italic;"># Customize x-axis text properties</span>    
        axis.text.x = element_text(family = <span style="color: #04cc04; font-style: italic;">"Arial"</span>, colour = <span style="color: #04cc04; font-style: italic;">"black"</span>, size = <span style="color: #cf2139; font-style: italic;">24</span>),
        <span style="color: #999999; font-style: italic;"># Customize y-axis text properties</span>    
        axis.text.y = element_text(family = <span style="color: #04cc04; font-style: italic;">"Arial"</span>, colour = <span style="color: #04cc04; font-style: italic;">"black"</span>, size = <span style="color: #cf2139; font-style: italic;">24</span>),
        <span style="color: #999999; font-style: italic;"># Customize subtitle text properties</span>    
        plot.subtitle = element_text(family = <span style="color: #04cc04; font-style: italic;">"Arial"</span>, size = <span style="color: #cf2139; font-style: italic;">24</span>, colour = <span style="color: #04cc04; font-style: italic;">"black"</span>, hjust = <span style="color: #cf2139; font-style: italic;">0.5</span>),
        <span style="color: #999999; font-style: italic;"># Customize y-axis title properties</span>    
        axis.title.y = element_text(family = <span style="color: #04cc04; font-style: italic;">"Arial"</span>, size = <span style="color: #cf2139; font-style: italic;">24</span>, angle = <span style="color: #cf2139; font-style: italic;">90</span>),
        <span style="color: #999999; font-style: italic;"># Customize x-axis title properties </span>   
        axis.title.x = element_text(family = <span style="color: #04cc04; font-style: italic;">"Arial"</span>, size = <span style="color: #cf2139; font-style: italic;">24</span>, angle = <span style="color: #cf2139; font-style: italic;">00</span>),
        <span style="color: #999999; font-style: italic;"># Customize legend text properties</span>    
        legend.text = element_text(size = <span style="color: #cf2139; font-style: italic;">10</span>, family = <span style="color: #04cc04; font-style: italic;">"Arial"</span>), 
        <span style="color: #999999; font-style: italic;"># Customize legend title properties</span>    
        legend.title = element_text(size = <span style="color: #cf2139; font-style: italic;">20</span>, family = <span style="color: #04cc04; font-style: italic;">"Arial"</span>)
      ) + 
      <span style="color: #999999; font-style: italic;"># Set the subtitle of the plot</span>
      labs(subtitle = paste0(<span style="color: #04cc04; font-style: italic;">"Volcano plot-"</span>, projname))
  }
  
  <span style="color: #999999; font-style: italic;"><strong># Calculate the number of up-regulated and down-regulated genes:</strong></span>
  <span style="color: #999999; font-style: italic;"># Write the up-regulated and down-regulated genes to a TSV file</span>
  write_tsv(rbind(volcano_up, volcano_down), paste0(projname, <span style="color: #04cc04; font-style: italic;">"/"</span>, <span style="color: #04cc04; font-style: italic;">"all_top_down.tsv"</span>))
  
  <span style="color: #999999; font-style: italic;"># Create the top table containing the top up-regulated and down-regulated genes</span>
  top <- rbind(volcano_up[<span style="color: #cf2139; font-style: italic;">1</span>:ntop, ], volcano_down[<span style="color: #cf2139; font-style: italic;">1</span>:ntop, ])
  <span style="color: #999999; font-style: italic;"># Reorder columns in the top table</span>
  top <- top[, <span style="color: #f79502; font-style: italic;">c</span>(<span style="color: #cf2139; font-style: italic;">1</span>:<span style="color: #cf2139; font-style: italic;">6</span>, <span style="color: #cf2139; font-style: italic;">11</span>, <span style="color: #cf2139; font-style: italic;">7</span>, <span style="color: #cf2139; font-style: italic;">8</span>, <span style="color: #cf2139; font-style: italic;">9</span>, <span style="color: #cf2139; font-style: italic;">10</span>, <span style="color: #cf2139; font-style: italic;">12</span>)]
  <span style="color: #999999; font-style: italic;"># Rename the 7th column as "n.log10(adj.P.Val)"</span>
  colnames(top)[<span style="color: #cf2139; font-style: italic;">7</span>] <- <span style="color: #04cc04; font-style: italic;">"n.log10(adj.P.Val)"</span>
  
  <span style="color: #999999; font-style: italic;"># Save the top table as a CSV file in the project folder</span>
  write.csv(top, paste0(projname, <span style="color: #04cc04; font-style: italic;">"/"</span>, <span style="color: #04cc04; font-style: italic;">"top.csv"</span>), <span style="color: #f79502; font-style: italic;">quote</span> = <span style="color: #0548e8; font-style: italic;">FALSE</span>, row.names = <span style="color: #0548e8; font-style: italic;">FALSE</span>)
  
  <span style="color: #999999; font-style: italic;"># Display the number of up-regulated and down-regulated genes</span>
  message(sprintf(<span style="color: #04cc04; font-style: italic;">"Up-regulated genes: %s; Down-regulated genes: %s"</span>, nrow(volcano_up), nrow(volcano_down)))
  
  <span style="color: #999999; font-style: italic;"># Return the generated plot</span>
  <span style="color: #f79502; font-style: italic;">return</span>(p)
}


<span style="color: #999999; font-style: italic;"># Function to creates a Venn diagram to visualize common differentially expressed genes among specified GEO datasets.</span>
MakeVenna <- <span style="color: #0548e8; font-style: italic;">function</span>(common, ntop = <span style="color: #cf2139; font-style: italic;">10</span>) {
  
  <span style="color: #999999; font-style: italic;"># Create a directory to store the output files</span>
  dir.create(<span style="color: #04cc04; font-style: italic;">"common"</span>, showWarnings = <span style="color: #0548e8; font-style: italic;">FALSE</span>)
  
  <span style="color: #999999; font-style: italic;"># Reading all all_top_down files and processing them</span>
  com_table <- <span style="color: #f79502; font-style: italic;">list</span>()
  
  <span style="color: #999999; font-style: italic;"># Process each dataset</span>
  <span style="color: #0548e8; font-style: italic;">for</span> (i <span style="color: #0548e8; font-style: italic;">in</span> <span style="color: #f79502; font-style: italic;">seq_along</span>(common)) {
    <span style="color: #999999; font-style: italic;"># Read the all_top_down file for the current dataset</span>
    DEG_table <- data.frame(read_tsv(paste0(common[i], <span style="color: #04cc04; font-style: italic;">"/"</span>, <span style="color: #04cc04; font-style: italic;">"all_top_down.tsv"</span>)))
    
    <span style="color: #999999; font-style: italic;"># Finding duplicated genes</span>
    <span style="color: #999999; font-style: italic;"># Find the indices of duplicated genes</span>
    dup_genes_n <- which(duplicated(DEG_table$Gene.symbol))
    <span style="color: #999999; font-style: italic;"># Get unique duplicated genes</span>
    dup_uniq_genes <- unique(DEG_table[dup_genes_n, <span style="color: #04cc04; font-style: italic;">"Gene.symbol"</span>])
    <span style="color: #999999; font-style: italic;"># Find the indices of all duplicated genes</span>
    find_dup <- which(DEG_table$Gene.symbol %in% dup_uniq_genes)
    <span style="color: #999999; font-style: italic;"># Extract all duplicated genes</span>
    all_dup <- DEG_table[find_dup, ]
    <span style="color: #999999; font-style: italic;"># Remove duplicated genes from the DEG table</span>
    DEG_table <- DEG_table[-find_dup, <span style="color: #cf2139; font-style: italic;">-11</span>]
    
    <span style="color: #999999; font-style: italic;"># Calculation the mean for logFC/AveExpr/t value/P-value/adjusted P-value/B value of duplicated genes</span>
    uniq_genes_list <- <span style="color: #f79502; font-style: italic;">list</span>()
    
    <span style="color: #999999; font-style: italic;"># Process each duplicated gene</span>
    <span style="color: #999999; font-style: italic;">for</span> (j <span style="color: #0548e8; font-style: italic;">in</span> <span style="color: #f79502; font-style: italic;">seq_along</span>(dup_uniq_genes)) {
      <span style="color: #999999; font-style: italic;"># Find the indices of the current duplicated gene</span>
      c <- which(all_dup$Gene.symbol %in% dup_uniq_genes[j])
      <span style="color: #999999; font-style: italic;"># Calculate the mean logFC</span>
      b <- <span style="color: #f79502; font-style: italic;">c</span>(mean(all_dup$logFC[<span style="color: #f79502; font-style: italic;">c</span>]),
             <span style="color: #999999; font-style: italic;"># Calculate the mean AveExpr</span>
             mean(all_dup$AveExpr[<span style="color: #f79502; font-style: italic;">c</span>]),
             <span style="color: #999999; font-style: italic;"># Calculate the mean t value</span>
             mean(all_dup$t[<span style="color: #f79502; font-style: italic;">c</span>]),
             <span style="color: #999999; font-style: italic;"># Calculate the mean P-value</span>
             mean(all_dup$P.Value[<span style="color: #f79502; font-style: italic;">c</span>]),
             <span style="color: #999999; font-style: italic;"># Calculate the mean adjusted P-value</span>
             mean(all_dup$adj.P.Val[<span style="color: #f79502; font-style: italic;">c</span>]),
             <span style="color: #999999; font-style: italic;"># Calculate the mean B value</span>
             mean(all_dup$B[<span style="color: #f79502; font-style: italic;">c</span>]))
      
      <span style="color: #999999; font-style: italic;"># Assign column names and create a new data frame with the mean values</span>
      <span style="color: #999999; font-style: italic;"># Assign column names to the mean values</span>
     <span style="color: #f79502; font-style: italic;">names</span>(b) <- <span style="color: #f79502; font-style: italic;">c</span>(<span style="color: #04cc04; font-style: italic;">"logFC"</span>, <span style="color: #04cc04; font-style: italic;">"AveExpr"</span>, <span style="color: #04cc04; font-style: italic;">"t"</span>, <span style="color: #04cc04; font-style: italic;">"P.Value"</span>, <span style="color: #04cc04; font-style: italic;">"adj.P.Val"</span>, <span style="color: #04cc04; font-style: italic;">"B"</span>)
      <span style="color: #999999; font-style: italic;"># Create a new data frame with the mean values</span>
      d <- data.frame(<span style="color: #f79502; font-style: italic;">c</span>(all_dup[c[<span style="color: #cf2139; font-style: italic;">1</span>], -<span style="color: #f79502; font-style: italic;">c</span>(<span style="color: #cf2139; font-style: italic;">2</span>, <span style="color: #cf2139; font-style: italic;">3</span>, <span style="color: #cf2139; font-style: italic;">4</span>, <span style="color: #cf2139; font-style: italic;">5</span>, <span style="color: #cf2139; font-style: italic;">6</span>, <span style="color: #cf2139; font-style: italic;">7</span>, <span style="color: #cf2139; font-style: italic;">11</span>)], b))
      <span style="color: #999999; font-style: italic;"># Reorder the columns of the data frame</span>
      d <- d[, <span style="color: #f79502; font-style: italic;">c</span>(<span style="color: #cf2139; font-style: italic;">1</span>, <span style="color: #cf2139; font-style: italic;">6</span>, <span style="color: #cf2139; font-style: italic;">7</span>, <span style="color: #cf2139; font-style: italic;">8</span>, <span style="color: #cf2139; font-style: italic;">9</span>, <span style="color: #cf2139; font-style: italic;">10</span>, <span style="color: #cf2139; font-style: italic;">11</span>, <span style="color: #cf2139; font-style: italic;">2</span>, <span style="color: #cf2139; font-style: italic;">3</span>, <span style="color: #cf2139; font-style: italic;">4</span>, <span style="color: #cf2139; font-style: italic;">5</span>)]
      <span style="color: #999999; font-style: italic;"># Store the data frame in the list for unique genes</span>
      uniq_genes_list[[j]] <- d
    }
    
    <span style="color: #999999; font-style: italic;"># Joining the mean calculated and unique genes with the rest of genes</span>
    <span style="color: #999999; font-style: italic;"># Combine the data frames in the list into one data frame</span>
    uniq_genes_df <- do.call(rbind, uniq_genes_list)
    <span style="color: #999999; font-style: italic;"># Append the unique genes data frame to the DEG_table</span>
    DEG_table <- rbind(DEG_table, uniq_genes_df)
    
    <span style="color: #999999; font-style: italic;"># Store the processed DEG table for the current dataset</span>
    com_table[[common[i]]] <- DEG_table
  }
  
  <span style="color: #999999; font-style: italic;"># Calculating up and down-regulated genes for each dataset</span>
  <span style="color: #999999; font-style: italic;"># Create an empty list to store the up and down-regulated genes</span>
  DEGslist <- <span style="color: #f79502; font-style: italic;">list</span>()
  
  <span style="color: #0548e8; font-style: italic;">for</span> (i <span style="color: #0548e8; font-style: italic;">in</span> <span style="color: #f79502; font-style: italic;">seq_along</span>(com_table)) {
    <span style="color: #999999; font-style: italic;"># Extract the 8th column (DEGs) and store it in the list</span>
    DEGslist[[common[i]]] <- com_table[[i]][<span style="color: #cf2139; font-style: italic;">8</span>]
  }
  
  <span style="color: #999999; font-style: italic;"># Find common genes among all datasets</span>
  com_prob <- Reduce(intersect, DEGslist)
  
  <span style="color: #999999; font-style: italic;"># Writing all common genes into separated files as well as the top files</span>
  <span style="color: #999999; font-style: italic;"># Create an empty list to store all common genes</span>
  all_com <- <span style="color: #f79502; font-style: italic;">list</span>()
  <span style="color: #999999; font-style: italic;"># Create an empty list to store all common genes (including all samples)</span>
  all_com_all <- <span style="color: #f79502; font-style: italic;">list</span>()
  
  <span style="color: #0548e8; font-style: italic;">for</span> (i <span style="color: #0548e8; font-style: italic;">in</span> <span style="color: #f79502; font-style: italic;">seq_along</span>(com_table)) {
    <span style="color: #999999; font-style: italic;"># Extract the common differentially expressed genes for the current dataset</span>
    comn <- which(com_table[[<span style="color: #f79502; font-style: italic;">names</span>(com_table)[i]]][[<span style="color: #04cc04; font-style: italic;">"Gene.symbol"</span>]] %in% com_prob$Gene.symbol)
    <span style="color: #999999; font-style: italic;"># Subset the current dataset with the common DEGs</span>
    com_final_DEG  <- com_table[[<span style="color: #f79502; font-style: italic;">names</span>(com_table)[i]]][comn, ]
    <span style="color: #999999; font-style: italic;"># Sort the subsetted data frame based on the "DEG" column</span>
    com_final_DEG <- com_final_DEG[order(com_final_DEG$DEG), ]
    
    <span style="color: #999999; font-style: italic;"># Save the common differentially expressed genes as a TSV file</span>
    write_tsv(com_final_DEG, paste0(<span style="color: #04cc04; font-style: italic;">"common"</span>, <span style="color: #04cc04; font-style: italic;">"/"</span>,<span style="color: #f79502; font-style: italic;">names</span>(com_table)[i], <span style="color: #04cc04; font-style: italic;">"_common_DEGs.tsv"</span>))
    
    <span style="color: #999999; font-style: italic;"># Extract the top up-regulated genes for the current dataset</span>
    com_final_up <- com_final_DEG[which(com_final_DEG$DEG == <span style="color: #04cc04; font-style: italic;">"Upregulated"</span>), ]
    <span style="color: #999999; font-style: italic;"># Sort the subsetted data frame based on the "logFC" column in decreasing order</span>
    com_final_up <- com_final_up[order(com_final_up$logFC, decreasing = <span style="color: #0548e8; font-style: italic;">TRUE</span>), ]
    
    <span style="color: #999999; font-style: italic;"># Extract the top down-regulated genes for the current dataset</span>
    com_final_down <- com_final_DEG[which(com_final_DEG$DEG == <span style="color: #04cc04; font-style: italic;">"Downregulated"</span>), ]
    <span style="color: #999999; font-style: italic;"># Sort the subsetted data frame based on the "logFC" column in increasing order</span>
    com_final_down <- com_final_down[order(com_final_down$logFC, decreasing = <span style="color: #0548e8; font-style: italic;">FALSE</span>), ]
    
    <span style="color: #999999; font-style: italic;"># Save the top common differentially expressed genes as a TSV file</span>
    write_tsv(rbind(com_final_up[<span style="color: #cf2139; font-style: italic;">1</span>:ntop, ], com_final_down[<span style="color: #cf2139; font-style: italic;">1</span>:ntop, ]), paste0(<span style="color: #04cc04; font-style: italic;">"common"</span>, <span style="color: #04cc04; font-style: italic;">"/"</span>,<span style="color: #f79502; font-style: italic;">names</span>(com_table)[i], <span style="color: #04cc04; font-style: italic;">"_top_common_DEGs.tsv"</span>))
    
    <span style="color: #999999; font-style: italic;"># Store the common differentially expressed genes and all genes for each dataset</span>
    <span style="color: #999999; font-style: italic;"># Store the subsetted com_final_DEG data frame in all_com list</span>
    all_com[[<span style="color: #f79502; font-style: italic;">names</span>(com_table)[i]]] <- com_final_DEG[, <span style="color: #f79502; font-style: italic;">c</span>(<span style="color: #cf2139; font-style: italic;">2</span>:<span style="color: #cf2139; font-style: italic;">7</span>)]
    <span style="color: #999999; font-style: italic;"># Store the complete com_final_DEG data frame in all_com_all list</span>
    all_com_all[[<span style="color: #f79502; font-style: italic;">names</span>(com_table)[i]]] <- com_final_DEG
  }
  
  <span style="color: #999999; font-style: italic;"># Calculate the mean of logFC for all common genes in each dataset</span> 
  <span style="color: #999999; font-style: italic;"># Combine the data frames in all_com list into a single data frame, all_com_df</span>
  all_com_df <- do.call(cbind, all_com)
  <span style="color: #999999; font-style: italic;"># Find the column indices of columns containing "logFC" in their names</span>
  logfc_n <- grep(<span style="color: #04cc04; font-style: italic;">"logFC"</span>, colnames(all_com_df))
  
  <span style="color: #999999; font-style: italic;"># Extract the columns containing logFC values from all_com_all_df</span>
  logfc_df <- all_com_df[, logfc_n]
  
  <span style="color: #999999; font-style: italic;"># Calculate the row means of logFC values to obtain the mean logFC for each common gene</span>
  logfc_df <- data.frame(logFC = rowMeans(logfc_df))
  
  <span style="color: #999999; font-style: italic;"># Extract columns containing adj.P.Val values from all_com_df:</span>
  <span style="color: #999999; font-style: italic;"># Find the column indices of columns containing "adj.P.Val" in their names</span>
  adj.P.Val_n <- grep(<span style="color: #04cc04; font-style: italic;">"adj.P.Val"</span>, colnames(all_com_df))
  <span style="color: #999999; font-style: italic;"># Extract the columns containing adj.P.Val values</span>
  adj.P.Val_df <- all_com_df[, adj.P.Val_n]
  <span style="color: #999999; font-style: italic;"># Calculate the row means of adj.P.Val values to obtain the mean adj.P.Val for each common gene:</span>
  <span style="color: #999999; font-style: italic;"># Create a data frame with a single column "adj.P.Val" containing the mean adj.P.Val values</span>
  adj.P.Val_df <- data.frame(adj.P.Val = rowMeans(adj.P.Val_df))
  
  <span style="color: #999999; font-style: italic;"># Extract columns containing B values from all_com_df</span>
  B_n <- grep(<span style="color: #04cc04; font-style: italic;">"B"</span>, colnames(all_com_df))
  B_df <- all_com_df[, B_n]
  <span style="color: #999999; font-style: italic;"># Calculate the row means of B values to obtain the mean B for each common gene</span>
  B_df <- data.frame(B = rowMeans(B_df))
  
  <span style="color: #999999; font-style: italic;"># Extract columns containing t values from all_com_df</span>
  t_n <- grep(<span style="color: #04cc04; font-style: italic;">".t"</span>, colnames(all_com_df))
  t_df <- all_com_df[, t_n]
  <span style="color: #999999; font-style: italic;"># Calculate the row means of t values to obtain the mean t for each common gene</span>
  t_df <- data.frame(t = rowMeans(t_df))
  
  <span style="color: #999999; font-style: italic;"># Extract columns containing P.Value values from all_com_df</span>
  P.Value_n <- grep(<span style="color: #04cc04; font-style: italic;">"P.Value"</span>, colnames(all_com_df))
  P.Value_df <- all_com_df[, P.Value_n]
  <span style="color: #999999; font-style: italic;"># Calculate the row means of P.Value values to obtain the mean P.Value for each common gene</span>
  P.Value_df <- data.frame(P.Value = rowMeans(P.Value_df))
  
  <span style="color: #999999; font-style: italic;"># Extract columns containing AveExpr values from all_com_df:</span>
  <span style="color: #999999; font-style: italic;"># Find the column indices of columns containing "AveExpr" in their names</span>
  AveExpr_n <- grep(<span style="color: #04cc04; font-style: italic;">"AveExpr"</span>, colnames(all_com_df))
  <span style="color: #999999; font-style: italic;"># Extract the columns containing AveExpr values</span>
  AveExpr_df <- all_com_df[, AveExpr_n]
  <span style="color: #999999; font-style: italic;"># Calculate the row means of AveExpr values to obtain the mean AveExpr for each common gene:</span>
  <span style="color: #999999; font-style: italic;"># Create a data frame with a single column "AveExpr" containing the mean AveExpr values</span>
  AveExpr_df <- data.frame(AveExpr = rowMeans(AveExpr_df))
  
  <span style="color: #999999; font-style: italic;"># Combining all calculated means into one data frame </span>
  all_com_all_df <- data.frame(do.call(cbind, all_com_all[[<span style="color: #cf2139; font-style: italic;">1</span>]]))
  <span style="color: #999999; font-style: italic;"># Removing unnecessary columns from all_com_all_df</span>
  all_com_all_df <- all_com_all_df[, -<span style="color: #f79502; font-style: italic;">c</span>(<span style="color: #cf2139; font-style: italic;">2</span>:<span style="color: #cf2139; font-style: italic;">7</span>)]
  <span style="color: #999999; font-style: italic;"># Combining the calculated means with all_com_all_df</span>
  all_com_all_df <- cbind(all_com_all_df, logfc_df, adj.P.Val_df, B_df, t_df, AveExpr_df, P.Value_df)
  <span style="color: #999999; font-style: italic;"># Reordering the columns of all_com_all_df</span>
  all_com_all_df <- all_com_all_df[, <span style="color: #f79502; font-style: italic;">c</span>(<span style="color: #cf2139; font-style: italic;">1</span>, <span style="color: #cf2139; font-style: italic;">6</span>, <span style="color: #cf2139; font-style: italic;">10</span>, <span style="color: #cf2139; font-style: italic;">9</span>, <span style="color: #cf2139; font-style: italic;">11</span>, <span style="color: #cf2139; font-style: italic;">7</span>, <span style="color: #cf2139; font-style: italic;">8</span>, <span style="color: #cf2139; font-style: italic;">2</span>, <span style="color: #cf2139; font-style: italic;">3</span>, <span style="color: #cf2139; font-style: italic;">4</span>, <span style="color: #cf2139; font-style: italic;">5</span>)]
  
  <span style="color: #999999; font-style: italic;"># Writing the combined data frame of all calculated means to a TSV file</span>
  write_tsv(all_com_all_df, paste0(<span style="color: #04cc04; font-style: italic;">"common"</span>, <span style="color: #04cc04; font-style: italic;">"/"</span>, <span style="color: #04cc04; font-style: italic;">"all_mean_common_DEGs.tsv"</span>))
  
  <span style="color: #999999; font-style: italic;"># Displaying a message to indicate that the common genes have been saved</span>
  message(<span style="color: #04cc04; font-style: italic;">"+++++ Common genes between all datasets have been saved into the 'common' folder. +++++"</span>)
  
  <span style="color: #999999; font-style: italic;"># Creating a list of DEGs for each dataset to generate a Venn diagram</span>
  venlist <- <span style="color: #f79502; font-style: italic;">list</span>()
  <span style="color: #999999; font-style: italic;"># Iterating over the DEGslist to populate the venlist</span>
  <span style="color: #0548e8; font-style: italic;">for</span> (i <span style="color: #0548e8; font-style: italic;">in</span> <span style="color: #f79502; font-style: italic;">seq_along</span>(DEGslist)) {
    venlist[[common[i]]] <- DEGslist[[i]][[<span style="color: #cf2139; font-style: italic;">1</span>]]
  }
  
  <span style="color: #999999; font-style: italic;"># Generate the Venn diagram using the logFC mean values for common genes</span>
  <span style="color: #999999; font-style: italic;"># Set the seed for random number generation</span>
  set.seed(<span style="color: #cf2139; font-style: italic;">1234</span>)
  <span style="color: #999999; font-style: italic;"># Define color palette for the Venn diagram</span>
  my_pal <- <span style="color: #f79502; font-style: italic;">c</span>(<span style="color: #1B9E77; font-style: italic;">"#1B9E77"</span>, <span style="color: #7570B3; font-style: italic;">"#7570B3"</span>, <span style="color: #E7298A; font-style: italic;">"#E7298A"</span>, <span style="color: #; font-style: italic;">"#66A61E"</span>, <span style="color: #E6AB02; font-style: italic;">"#E6AB02"</span>, <span style="color: #A6761D; font-style: italic;">"#A6761D"</span>, <span style="color: #666666; font-style: italic;">"#666666"</span>, <span style="color: #9A7D0A; font-style: italic;">"#9A7D0A"</span>)
  <span style="color: #999999; font-style: italic;"># Create the Venn diagram using ggvenn package</span>
  venaa <- ggvenn(venlist, 
                  fill_color = my_pal,
                  stroke_size = <span style="color: #cf2139; font-style: italic;">0.5</span>,
                  set_name_size = <span style="color: #cf2139; font-style: italic;">4</span>)
  
  <span style="color: #999999; font-style: italic;"># Return the generated Venn diagram</span>
  <span style="color: #f79502; font-style: italic;">return</span>(venaa)
}



<span style="color: #999999; font-style: italic;"># Function to prepares files for network analysis using the specified GEO datasets.</span>
Makenetworkanalyst <- <span style="color: #0548e8; font-style: italic;">function</span>(projname,
                               compare = <span style="color: #04cc04; font-style: italic;">"grade"</span>,
                               groups) {
  <span style="color: #999999; font-style: italic;"># Reading files</span>
  <span style="color: #999999; font-style: italic;"># Create 'networkanalyst' directory to store the output files</span>
  dir.create(<span style="color: #04cc04; font-style: italic;">"networkanalyst"</span>, showWarnings = <span style="color: #0548e8; font-style: italic;">FALSE</span>)
  
  <span style="color: #999999; font-style: italic;"># Read the 'matrix.csv' file</span>
  matrix <- read_csv(paste0(projname, <span style="color: #04cc04; font-style: italic;">"/"</span>, <span style="color: #04cc04; font-style: italic;">"matrix.csv"</span>))
  matrix <- data.frame(matrix)
  
  <span style="color: #999999; font-style: italic;"># Read the 'phenotype.tsv' file</span>
  phenotype <- read_tsv(paste0(projname, <span style="color: #04cc04; font-style: italic;">"/"</span>, <span style="color: #04cc04; font-style: italic;">"phenotype.tsv"</span>))
  phenotype <- data.frame(phenotype)
  
  <span style="color: #999999; font-style: italic;"># Set the row names of the 'phenotype' dataframe</span>
  rownames(phenotype) <- phenotype[, <span style="color: #cf2139; font-style: italic;">1</span>]
  
  <span style="color: #999999; font-style: italic;"># Calculating DEGs:</span>
  <span style="color: #999999; font-style: italic;"># Read the 'all_top_down.tsv' file containing up and down genes</span>
  updown <- data.frame(read_tsv(paste0(projname, <span style="color: #04cc04; font-style: italic;">"/"</span>, <span style="color: #04cc04; font-style: italic;">"all_top_down.tsv"</span>)))
  
  <span style="color: #999999; font-style: italic;"># Creating a file that contains up and down genes based on groups specified:</span>
  
  <span style="color: #999999; font-style: italic;"># Identify the rows in 'matrix' corresponding to up and down genes</span>
  updown_genes <- which(matrix$ID %in% updown$ID)
  <span style="color: #999999; font-style: italic;"># Create a new matrix 'updown_matrix' containing the identified rows</span>
  updown_matrix <- matrix[updown_genes, ]
 
  <span style="color: #999999; font-style: italic;"># Add gene symbols to the 'updown_matrix' dataframe</span>
  updown_matrix <- cbind(updown$Gene.symbol, updown_matrix)
  <span style="color: #999999; font-style: italic;"># Rename the first column as "Symbol"</span>
  colnames(updown_matrix)[<span style="color: #cf2139; font-style: italic;">1</span>] <-  <span style="color: #04cc04; font-style: italic;">"Symbol" </span>
  <span style="color: #999999; font-style: italic;"># Remove the second column from the dataframe</span>
  updown_matrix <- updown_matrix[, <span style="color: #cf2139; font-style: italic;">-2</span>]
  
  <span style="color: #999999; font-style: italic;"># Extract the gene symbols from the 'updown_matrix' dataframe</span>
  genes <- updown_matrix$Symbol
  <span style="color: #999999; font-style: italic;"># Remove the first column (gene symbols) from 'updown_matrix', round the remaining expression values to 2 decimal places</span>
  updown_matrix <- <span style="color: #f79502; font-style: italic;">round</span>(updown_matrix[<span style="color: #cf2139; font-style: italic;">-1</span>], <span style="color: #cf2139; font-style: italic;">2</span>)
  <span style="color: #999999; font-style: italic;"># Add the gene symbols column back to the 'updown_matrix' dataframe</span>
  updown_matrix <- cbind(genes, updown_matrix)
  
  <span style="color: #0548e8; font-style: italic;">if</span> (compare ==  <span style="color: #04cc04; font-style: italic;">"subtype"</span>) {
    <span style="color: #999999; font-style: italic;"># Prepare data based on subtype comparison</span>
    
    <span style="color: #999999; font-style: italic;"># Find the column index for 'pam50' in 'phenotype'</span>
    phecoln <- grep( <span style="color: #04cc04; font-style: italic;">"pam50"</span>, phenotype)
    
    <span style="color: #999999; font-style: italic;"># Check if the length of 'phecoln' is greater than 0 (if there are subtype columns present in 'phenotype' dataframe.)</span>
    <span style="color: #0548e8; font-style: italic;">if</span> (<span style="color: #f79502; font-style: italic;">length</span>(phecoln) > <span style="color: #cf2139; font-style: italic;">0</span>) {
      <span style="color: #999999; font-style: italic;"># Extract the subtype column from 'phenotype'</span>
      pheno <- data.frame(phenotype[, phecoln])
      <span style="color: #999999; font-style: italic;"># Rename the column to "subtype"</span>
      colnames(pheno) <-  <span style="color: #04cc04; font-style: italic;">"subtype"</span>
      
      <span style="color: #999999; font-style: italic;"># Clean up the subtype values by removing the "pam50" prefix</span>
      pheno$subtype <- gsub( <span style="color: #04cc04; font-style: italic;">"pam50.+:"</span>,  <span style="color: #04cc04; font-style: italic;">""</span>, pheno$subtype)
      <span style="color: #999999; font-style: italic;"># Assign the row names of 'phenotype' as the row names of 'pheno'</span>
      rownames(pheno) <- rownames(phenotype)
      
      <span style="color: #999999; font-style: italic;"># Creating groups of samples based on specified conditions:</span>
      <span style="color: #999999; font-style: italic;"># Create an empty list to store groups of samples</span>
      grouplis <- <span style="color: #f79502; font-style: italic;">list</span>()
      <span style="color: #0548e8; font-style: italic;">for</span> (i <span style="color: #0548e8; font-style: italic;">in</span> <span style="color: #f79502; font-style: italic;">seq_along</span>(groups)) {
        <span style="color: #999999; font-style: italic;"># Combine the group names using the '|' separator</span>
        name <- paste0(groups[[i]], collapse = <span style="color: #04cc04; font-style: italic;">"|"</span>)
        <span style="color: #999999; font-style: italic;"># Find the rows in 'pheno$subtype' that match the group name</span>
        rownum <- grep(name, pheno$subtype)
        <span style="color: #999999; font-style: italic;"># Store the matching row names in 'grouplis' with the group name as the key</span>
        grouplis[[<span style="color: #f79502; font-style: italic;">names</span>(groups)[i]]] <- rownames(pheno)[rownum]
      }
    }
  } <span style="color: #0548e8; font-style: italic;">else</span> {
    <span style="color: #999999; font-style: italic;"># Preparing the breast cancer grades into a dataframe:</span>
    <span style="color: #999999; font-style: italic;"># Find the column index containing "grade:" in the phenotype data</span>
    gradecoln <- grep( <span style="color: #04cc04; font-style: italic;">"grade:"</span>, phenotype)
    <span style="color: #999999; font-style: italic;"># Check if any column contains "grade:"</span>
    <span style="color: #0548e8; font-style: italic;">if</span> (<span style="color: #f79502; font-style: italic;">length</span>(gradecoln) > <span style="color: #cf2139; font-style: italic;">0</span>) {
      <span style="color: #999999; font-style: italic;"># Extract the column with "grade:" and create a data frame</span>
      grade <- data.frame(phenotype[, gradecoln])
      <span style="color: #999999; font-style: italic;"># Rename the column to "grade"</span>
      colnames(grade) <-  <span style="color: #04cc04; font-style: italic;">"grade"</span>
      <span style="color: #999999; font-style: italic;"># Remove "grade:", "grade: ", and "B-R grade: " from the values in the column</span>
      grade$grade <- gsub( <span style="color: #04cc04; font-style: italic;">"grade:|grade: |B-R grade: "</span>,  <span style="color: #04cc04; font-style: italic;">""</span>, grade$grade)
      <span style="color: #999999; font-style: italic;"># Remove any characters after "=" in the values</span>
      grade$grade <- gsub( <span style="color: #04cc04; font-style: italic;">"=.+"</span>,  <span style="color: #04cc04; font-style: italic;">""</span>, grade$grade)
      <span style="color: #999999; font-style: italic;"># Find the rows with "III" in the grade column</span>
      greek <- grep( <span style="color: #04cc04; font-style: italic;">"III"</span>, grade$grade)
      <span style="color: #999999; font-style: italic;"># Check if any rows contain "III"</span>
      <span style="color: #0548e8; font-style: italic;">if</span> (<span style="color: #f79502; font-style: italic;">length</span>(greek) > <span style="color: #cf2139; font-style: italic;">0</span>) {
        <span style="color: #999999; font-style: italic;"># Find the rows with "I", "II", "III" in the grade column</span>
        one <- which(factor(grade$grade) ==  <span style="color: #04cc04; font-style: italic;">"I"</span>)
        two <- which(factor(grade$grade) ==  <span style="color: #04cc04; font-style: italic;">"II"</span>)
        three <- which(factor(grade$grade) ==  <span style="color: #04cc04; font-style: italic;">"III"</span>)
        <span style="color: #999999; font-style: italic;"># Replace the rows with "I" with "1"</span>
        grade[one, ] <-  <span style="color: #04cc04; font-style: italic;">"1"</span>
        <span style="color: #999999; font-style: italic;"># Replace the rows with "II" with "2"</span>
        grade[two, ] <-  <span style="color: #04cc04; font-style: italic;">"2"</span>
        <span style="color: #999999; font-style: italic;"># Replace the rows with "III" with "3"</span>
        grade[three, ] <-  <span style="color: #04cc04; font-style: italic;">"3"</span>
      }
      <span style="color: #999999; font-style: italic;"># Set the row names of the 'grade' data frame to match the row names of the 'phenotype' data frame</span>
      rownames(grade) <- rownames(phenotype)
      <span style="color: #999999; font-style: italic;"># Create an empty list to store the group information</span>
      grouplis <- <span style="color: #f79502; font-style: italic;">list</span>()
      <span style="color: #0548e8; font-style: italic;">for</span> (i <span style="color: #0548e8; font-style: italic;">in</span> <span style="color: #f79502; font-style: italic;">seq_along</span>(groups)) {
        <span style="color: #999999; font-style: italic;"># Concatenate the elements of 'groups' with a '|' separator</span>
        name <- paste0(groups[[i]], collapse = <span style="color: #04cc04; font-style: italic;">"|"</span>)
        <span style="color: #999999; font-style: italic;"># Find the rows in the 'grade' data frame where the 'grade' column matches the group name</span>
        rownum <- grep(name, grade$grade)
        <span style="color: #999999; font-style: italic;"># Store the row names corresponding to the group in the 'grouplis' list</span>
        grouplis[[<span style="color: #f79502; font-style: italic;">names</span>(groups)[i]]] <- rownames(grade)[rownum]
      }
    }
  }
  
  <span style="color: #999999; font-style: italic;"># Renaming the matrix file according to the groups name specified by the user:</span>
  <span style="color: #999999; font-style: italic;"># Bind the column names of 'updown_matrix' as the first row of 'updown_matrix'</span>
  updown_matrix <- rbind(colnames(updown_matrix), updown_matrix)
  <span style="color: #0548e8; font-style: italic;">for</span> (i <span style="color: #0548e8; font-style: italic;">in</span> <span style="color: #f79502; font-style: italic;">seq_along</span>(grouplis)) {
    <span style="color: #999999; font-style: italic;"># Concatenate the elements of 'grouplis[[i]]' with a '|' separator</span>
    matname <- paste0(grouplis[[i]], collapse = <span style="color: #04cc04; font-style: italic;">"|"</span>)
    <span style="color: #999999; font-style: italic;"># Find the columns in 'updown_matrix' that match the group name specified by 'matname'</span>
    matcoln <- grep(matname, colnames(updown_matrix))
    <span style="color: #999999; font-style: italic;"># Replace the corresponding columns in the first row of 'updown_matrix' with the group name</span>
    updown_matrix[<span style="color: #cf2139; font-style: italic;">1</span>, matcoln] <-<span style="color: #f79502; font-style: italic;">names</span>(grouplis)[i]
  }
  
  <span style="color: #999999; font-style: italic;"># Adding column names and the second row according to the file for networkanalyst:</span>
  <span style="color: #999999; font-style: italic;"># Find the columns in the first row of 'updown_matrix' that match the names of 'grouplis'</span>
  changedcol <- grep(paste0(<span style="color: #f79502; font-style: italic;">names</span>(grouplis), collapse = <span style="color: #04cc04; font-style: italic;">"|"</span>), updown_matrix[<span style="color: #cf2139; font-style: italic;">1</span>, ])
  <span style="color: #999999; font-style: italic;"># Keep the first column and the columns selected by 'changedcol' in 'updown_matrix'</span>
  updown_matrix <- updown_matrix[, <span style="color: #f79502; font-style: italic;">c</span>(<span style="color: #cf2139; font-style: italic;">1</span>, changedcol)]
  <span style="color: #999999; font-style: italic;"># Replace the first cell of 'updown_matrix' with "#CLASS"</span>
  updown_matrix[<span style="color: #cf2139; font-style: italic;">1</span>, <span style="color: #cf2139; font-style: italic;">1</span>] <-  <span style="color: #04cc04; font-style: italic;">"#CLASS"</span>
  <span style="color: #999999; font-style: italic;"># Replace the column name of the first column in 'updown_matrix' with "#NAME"</span>
  colnames(updown_matrix)[<span style="color: #cf2139; font-style: italic;">1</span>] <-  <span style="color: #04cc04; font-style: italic;">"#NAME"</span>
  
  <span style="color: #999999; font-style: italic;"># Write 'updown_matrix' to a tab-separated file</span>
  write.table(updown_matrix, paste0( <span style="color: #04cc04; font-style: italic;">"networkanalyst"</span>, <span style="color: #04cc04; font-style: italic;">"/"</span>, projname,  <span style="color: #04cc04; font-style: italic;">"_dataanlys.txt"</span>), <span style="color: #f79502; font-style: italic;">quote</span> = <span style="color: #0548e8; font-style: italic;">FALSE</span>, row.names = <span style="color: #0548e8; font-style: italic;">FALSE</span>, sep = <span style="color: #f79502; font-style: italic;">"\t"</span>)
  
  <span style="color: #999999; font-style: italic;"># Display a message indicating that the file has been saved</span>
  message(<span style="color: #04cc04; font-style: italic;">"+++++ Your file has been saved in the networkanalyst folder. +++++"</span>)
}


	  
<span style="color: #c20818; font-style: italic;"><strong>#===================================================== Q1 =====================================================#</strong></span>
# other adjust methods ---> "holm", "hochberg", "hommel", "bonferroni", "BH", "BY", "fdr", "none"

<span style="color: #75020c; font-style: italic;"><strong>################################################# GSE25055 #################################################</strong></span> 
<span style="color: #999999; font-style: italic;"># Download the GEO dataset GSE25055</span>
DownloadGEO( <span style="color: #04cc04; font-style: italic;">"GSE25055"</span>)

<span style="color: #999999; font-style: italic;"># Perform DEG analysis on GSE25055 dataset</span>
DEGanalysis(projname = <span style="color: #04cc04; font-style: italic;">"GSE25055"</span>,
            compare = <span style="color: #04cc04; font-style: italic;">"grade"</span>,
            groups = <span style="color: #f79502; font-style: italic;">list</span>(normal = <span style="color: #f79502; font-style: italic;">c</span>(<span style="color: #04cc04; font-style: italic;">"1"</span>), tumor = <span style="color: #f79502; font-style: italic;">c</span>(<span style="color: #04cc04; font-style: italic;">"3"</span>)),
            adjust = <span style="color: #04cc04; font-style: italic;">"fdr"</span>)

# Create a volcano plot for GSE25055 dataset
volcano_GSE25055 <- Makevolcano(projname = <span style="color: #04cc04; font-style: italic;">"GSE25055"</span>,
                                padjlevel = <span style="color: #cf2139; font-style: italic;">0.05</span>,
                                uplogFC = <span style="color: #cf2139; font-style: italic;">1</span>,
                                downlogFC = <span style="color: #cf2139; font-style: italic;">-1</span>,
                                ntop = <span style="color: #cf2139; font-style: italic;">10</span>,
                                voltype = <span style="color: #04cc04; font-style: italic;">"ggplot"</span>)
volcano_GSE25055

<span style="color: #999999; font-style: italic;"># Prepare files for network analysis on GSE25055 dataset</span>
Makenetworkanalyst(projname = <span style="color: #04cc04; font-style: italic;">"GSE25055"</span>,
                   compare = <span style="color: #04cc04; font-style: italic;">"grade"</span>,
                   groups = <span style="color: #f79502; font-style: italic;">list</span>(normal = <span style="color: #f79502; font-style: italic;">c</span>(<span style="color: #04cc04; font-style: italic;">"1"</span>), tumor = <span style="color: #f79502; font-style: italic;">c</span>(<span style="color: #04cc04; font-style: italic;">"3"</span>)))

<span style="color: #75020c; font-style: italic;"><strong>################################################# GSE7390 #################################################</strong></span>
<span style="color: #999999; font-style: italic;"># Download the GEO dataset GSE7390</span>
DownloadGEO(<span style="color: #04cc04; font-style: italic;">"GSE7390"</span>)

<span style="color: #999999; font-style: italic;"># Perform DEG analysis on GSE7390 dataset</span>
DEGanalysis(projname = <span style="color: #04cc04; font-style: italic;">"GSE7390"</span>,
            compare = <span style="color: #04cc04; font-style: italic;">"grade"</span>,
            groups = <span style="color: #f79502; font-style: italic;">list</span>(normal = <span style="color: #f79502; font-style: italic;">c</span>(<span style="color: #04cc04; font-style: italic;">"1"</span>), tumor = <span style="color: #f79502; font-style: italic;">c</span>(<span style="color: #04cc04; font-style: italic;">"3"</span>)),
            adjust = <span style="color: #04cc04; font-style: italic;">"fdr"</span>)

<span style="color: #999999; font-style: italic;"># Create a volcano plot for GSE7390 dataset</span>
volcano_GSE7390 <- Makevolcano(projname = <span style="color: #04cc04; font-style: italic;">"GSE7390"</span>,
                               padjlevel = <span style="color: #cf2139; font-style: italic;">0.05</span>,
                               uplogFC = <span style="color: #cf2139; font-style: italic;">1</span>,
                               downlogFC = <span style="color: #cf2139; font-style: italic;">-1</span>,
                               ntop = <span style="color: #cf2139; font-style: italic;">10</span>,
                               voltype = <span style="color: #04cc04; font-style: italic;">"ggplot"</span>)
volcano_GSE7390

<span style="color: #999999; font-style: italic;"># Prepare files for network analysis on GSE7390 dataset</span>
Makenetworkanalyst(projname = <span style="color: #04cc04; font-style: italic;">"GSE7390"</span>,
                   compare = <span style="color: #04cc04; font-style: italic;">"grade"</span>,
                   groups = <span style="color: #f79502; font-style: italic;">list</span>(normal = <span style="color: #f79502; font-style: italic;">c</span>(<span style="color: #04cc04; font-style: italic;">"1"</span>), tumor = <span style="color: #f79502; font-style: italic;">c</span>(<span style="color: #04cc04; font-style: italic;">"3"</span>)))

<span style="color: #75020c; font-style: italic;"><strong>################################################# GSE11121 #################################################</strong></span>
<span style="color: #999999; font-style: italic;"># Download the GEO dataset GSE11121</span>
DownloadGEO(<span style="color: #04cc04; font-style: italic;">"GSE11121"</span>)

<span style="color: #999999; font-style: italic;"># Perform DEG analysis on GSE11121 dataset</span>
DEGanalysis(projname = <span style="color: #04cc04; font-style: italic;">"GSE11121"</span>,
            compare = <span style="color: #04cc04; font-style: italic;">"grade"</span>,
            groups = <span style="color: #f79502; font-style: italic;">list</span>(normal = <span style="color: #f79502; font-style: italic;">c</span>(<span style="color: #04cc04; font-style: italic;">"1"</span>), tumor = <span style="color: #f79502; font-style: italic;">c</span>(<span style="color: #04cc04; font-style: italic;">"3"</span>)),
            adjust = <span style="color: #04cc04; font-style: italic;">"fdr"</span>)

<span style="color: #999999; font-style: italic;"># Create a volcano plot for GSE11121 dataset</span>
volcano_GSE11121 <- Makevolcano(projname = <span style="color: #04cc04; font-style: italic;">"GSE11121"</span>,
                                padjlevel = <span style="color: #cf2139; font-style: italic;">0.05</span>,
                                uplogFC = <span style="color: #cf2139; font-style: italic;">1</span>,
                                downlogFC = <span style="color: #cf2139; font-style: italic;">-1</span>,
                                ntop = <span style="color: #cf2139; font-style: italic;">10</span>,
                                voltype = <span style="color: #04cc04; font-style: italic;">"ggplot"</span>)
volcano_GSE11121

<span style="color: #999999; font-style: italic;"># Prepare files for network analysis on GSE11121 dataset</span>
Makenetworkanalyst(projname = <span style="color: #04cc04; font-style: italic;">"GSE11121"</span>,
                   compare = <span style="color: #04cc04; font-style: italic;">"grade"</span>,
                   groups = <span style="color: #f79502; font-style: italic;">list</span>(normal = <span style="color: #f79502; font-style: italic;">c</span>(<span style="color: #04cc04; font-style: italic;">"1"</span>), tumor = <span style="color: #f79502; font-style: italic;">c</span>(<span style="color: #04cc04; font-style: italic;">"3"</span>)))

<span style="color: #75020c; font-style: italic;"><strong>################################################# GSE25065 #################################################</strong></span>
<span style="color: #999999; font-style: italic;"></span># Download the GEO dataset GSE25065
DownloadGEO(<span style="color: #04cc04; font-style: italic;">"GSE25065"</span>)

<span style="color: #999999; font-style: italic;"># Perform DEG analysis on GSE25065 dataset</span>
DEGanalysis(projname = <span style="color: #04cc04; font-style: italic;">"GSE25065"</span>,
            compare = <span style="color: #04cc04; font-style: italic;">"grade"</span>,
            groups = <span style="color: #f79502; font-style: italic;">list</span>(normal = <span style="color: #f79502; font-style: italic;">c</span>(<span style="color: #04cc04; font-style: italic;">"1"</span>), tumor = <span style="color: #f79502; font-style: italic;">c</span>(<span style="color: #04cc04; font-style: italic;">"3"</span>)),
            adjust = <span style="color: #04cc04; font-style: italic;">"fdr"</span>)

<span style="color: #999999; font-style: italic;"># Create a volcano plot for GSE25065 dataset</span>
volcano_GSE25065 <- Makevolcano(projname = <span style="color: #04cc04; font-style: italic;">"GSE25065"</span>,
                                padjlevel = <span style="color: #cf2139; font-style: italic;">0.05</span>,
                                uplogFC = <span style="color: #cf2139; font-style: italic;">1</span>,
                                downlogFC = <span style="color: #cf2139; font-style: italic;">-1</span>,
                                ntop = <span style="color: #cf2139; font-style: italic;">10</span>,
                                voltype = <span style="color: #04cc04; font-style: italic;">"ggplot"</span>)
volcano_GSE25065

<span style="color: #999999; font-style: italic;"># Prepare files for network analysis on GSE25065 dataset</span>
Makenetworkanalyst(projname = <span style="color: #04cc04; font-style: italic;">"GSE25065"</span>,
                   compare = <span style="color: #04cc04; font-style: italic;">"grade"</span>,
                   groups = <span style="color: #f79502; font-style: italic;">list</span>(normal = <span style="color: #f79502; font-style: italic;">c</span>(<span style="color: #04cc04; font-style: italic;">"1"</span>), tumor = <span style="color: #f79502; font-style: italic;">c</span>(<span style="color: #04cc04; font-style: italic;">"3"</span>)))


<span style="color: #999999; font-style: italic;"># Create a Venn diagram for the common differentially expressed genes across the datasets</span>
MakeVenna(common = <span style="color: #f79502; font-style: italic;">c</span>(<span style="color: #04cc04; font-style: italic;">"GSE25055"</span>, <span style="color: #04cc04; font-style: italic;">"GSE7390"</span>, <span style="color: #04cc04; font-style: italic;">"GSE11121"</span>, <span style="color: #04cc04; font-style: italic;""GSE25065"></span>), ntop = <span style="color: #cf2139; font-style: italic;">10</span>)


			
			
			
			
			
		</code></pre>

		</code></pre>
	</section>

	<footer class="footer">
	    <p>&copy; 2023 Mohammad Reza Mohajeri. All rights reserved.</p>
	</footer>
</body>
</html>
