<!DOCTYPE html>
<html>
<head>
	<meta charset="UTF-8">
	<title>Sample 1</title>
	<style>
		/* Header  Styles  */
		.header {
		  background-color: #00FF00; /* primary code:  #f2f2f2 / current code: Electric Green */ 
		  padding-top: 5px; /* Increase or decrease the top padding value as desired */
		  padding-bottom: 5px; /* Increase or decrease the bottom padding value as desired */
		  text-align: center;
		  border-bottom: 0.7px solid black; /* Add this line to create a solid line */

		}
		/* 	Red: #FF0000
			Orange: #FFA500
			Bright Yellow: #FFFF00
			Vibrant Blue: #00BFFF
			Electric Green: #00FF00
			Hot Pink: #FF69B4 
			*/

		.header h1 {
			font-size: 40px; /* Adjust the font size as desired */
			color: ; #333; 
			font-weight: bold; /* Add this line to make the text bold */
		}

		/* Navbar Styles */
		.navbar {
			background-color: #e6f7ff ; /* #f2f2f2 primary color */
			padding: 10px;
			border-bottom: 1px solid black; /* Add this line to create a solid line */
	
		}

		.menu {
			list-style-type: none;
			margin: 0;
			padding: 0;
		}

		.menu li {
			display: inline-block;
			margin-right: 10px;
		}

		.menu li a {
			display: block;
			padding: 10px;
			background-color: #DAF7A6; /* green */
			color: darkblue; /* Change the color to dark blue */
			text-decoration: none;
			
		}
		
		.menu li a:hover {
			background-color: #FFC300; /* Gold */
			color: darkgreen !important; /* Change the text color to dark green */
			text-decoration: none;
		} 

		/* Sidebar Styles */
		.sidebar {
			width: 20%;
			height: 5000px; /* Increase the height as desired */
			background-color: #e6f7ff;
			padding: 10px;
			float: left;
			border-right: 1px solid black; /* This will visually separate the sidebar and content sections. */

		}

		.sidebar ul {
			list-style-type: none;
			padding: 0;
		}

		.sidebar li {
			margin-bottom: 5px;
		}

		
		.sidebar .heading {
			cursor: default;
			transition: background-color 0.3s, color 0.3s; /* Add transition for smooth hover effect */
		}
		
		.sidebar .heading:hover {
			background-color: #FFC300; /* Apply hover background color */
			color: #fff; /* Change the text color on hover */
		}

		.sidebar .subheadings {
			display: none;
			padding-left: 10px;
		}

		.sidebar .subheadings li {
			margin-bottom: 5px;
		}
		
	
		
		
				/* Content Styles */
		
		.content {
			  font-size: 20px;
			  width: 75%;
			  height: 40px; /* Set the initial height when the dropdown is closed */
			  padding: 10px;
			  float: left;
			  background-color: #e6f7ff;
			  color: #333;
			  cursor: pointer;
			  transition: background-color 0.3s, color 0.3s;
			  overflow: hidden; /* Add overflow property to hide extra content */
			}

			.content:hover {
			  height: auto; /* Set the height to auto on hover to expand the dropdown */
			  background-color: #95bfba;
			  color: #000000;
			}

			.content.active {
			  height: auto; /* Set the height to auto when the dropdown is open */
			  background-color: #0000FF;
			}

			.content h2 {
			  font-size: 18px;
			  font-size: 24px; /*  to increase the font size of the "Content 1" heading ;Increase the font size to your desired value */

			}

			.content ul {
			  font-size: 16px;
			  list-style-type: none;
			  padding-left: 20px;
			  display: none;
			}

			.content.active ul {
			  display: block; /* Display the dropdown content when the dropdown is open */
			}
		
	
}



	</style>
</head>
<body>
	<div class="header">
		<h1>Sample 1</h1>
	</div>

	<div class="navbar">
		<ul class="menu">
			<li><a href="#">Tab 1</a></li>
			<li><a href="#">Tab 2</a></li>
			<li><a href="#">Tab 3</a></li>
			<li><a href="#">Tab 4</a></li>
			<li><a href="#">Tab 5</a></li>
			<li><a href="#">Tab 6</a></li>
			<li><a href="#">Tab 7</a></li>
		</ul>
	</div>

	<div class="sidebar">
		<ul>
			<li>
				<h2 class="heading">Heading 1</h2>
				<ul class="subheadings">
					<li>Subheading 1</li>
					<li>Subheading 2</li>
					<li>Subheading 3</li>
					<li>Subheading 4</li>
				</ul>
			</li>
			<li>
				<h2 class="heading">Heading 2</h2>
				<ul class="subheadings">
					<li>Subheading 1</li>
					<li>Subheading 2</li>
					<li>Subheading 3</li>
					<li>Subheading 4</li>
				</ul>
			</li>
			<li>
				<h2 class="heading">Heading 3</h2>
				<ul class="subheadings">
					<li>Subheading 1</li>
					<li>Subheading 2</li>
					<li>Subheading 3</li>
					<li>Subheading 4</li>
				</ul>
			</li>
			<li>
				<h2 class="heading">Heading 4</h2>
				<ul class="subheadings">
					<li>Subheading 1</li>
					<li>Subheading 2</li>
					<li>Subheading 3</li>
					<li>Subheading 4</li>
				</ul>
			</li>
			<li>
				<h2 class="heading">Heading 5</h2>
				<ul class="subheadings">
					<li>Subheading 1</li>
					<li>Subheading 2</li>
					<li>Subheading 3</li>
					<li>Subheading 4</li>
				</ul>
			</li>
			<li>
				
				<h2 class="heading">Heading 6</h2>
				<ul class="subheadings">
					<li>Subheading 1</li>
					<li>Subheading 2</li>
					<li>Subheading 3</li>
					<li>Subheading 4</li>
				</ul>
			</li>
			<li>
				<h2 class="heading">Heading 7</h2>
				<ul class="subheadings">
					<li>Subheading 1</li>
					<li>Subheading 2</li>
					<li>Subheading 3</li>
					<li>Subheading 4</li>
				</ul>
			</li>
		</ul>
	</div>
	
	
	
	<!-- A Sample for Content with FIGURE & TABLE Hyperlink 
	<div class="content">
		<h2 class="heading">Content 1</h2>
		<p>
			TEEEEXXXXTTTTT BEFORE TABLE1

			<a href="table1.html">(table 1)</a>,

			TEEEEXXXXTTTTT FIGURE1

			<a href="figure1.html">(figure 1)</a>. 

			TEXTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTT
		</p>
	</div>
-->

	<div class="content">
		<h2 class="heading">Content 1</h2>
		<p>
			Read about breast cancer in women
			<a href="table1.html">(table 1)</a>,
			the most common type of cancer in the UK
			<a href="figure1.html">(figure 1)</a>. 


			2-	Methodology
After selecting datasets using the PubMed, data pre-processing and data integration using a self-customized R pipeline for meta-analysis, the meta- DEG results were visualized, and functional analysis was done with different methods. The overall workflow is summarized in Figure 1. The R pipeline is composed of three main stages for conducting the meta-analysis of gene expression data including “data pre-processing”, “merging datasets using REM model for getting DEGs”, and “functional analysis”. A part of functional analysis was done by this pipeline including the heatmap for top 25 up/down DEGs (sorted based on samples from grade I and III), and the GSEA analysis for each of the 4 datasets and the combined one as well. Other parts of the functional analysis were carried out using the NetworkAnalyst platform that is a comprehensive gene expression profiling platform. 

2-1- Selection of datasets using the PubMed
For doing this microarray meta-analysis, the PubMed was initially queried using two keywords of “breast cancer” and “gene expression” and 316 papers were found. This number was reduced to 20 based on the research criteria and through reading the titles, abstracts, and full texts, sequentially. Among the remainder, only papers with GSE accession numbers were kept and the required info were extracted. 
Next, only 4 independent datasets were selected for next analyses as the number of samples in these 4 datasets were enough. Therefore, in order to increase the accuracy, use of other datasets was avoided. The chosen datasets had all been carried out with the same platform (Affymetrix Human Genome U133A Array), with patient tissue samples taken before systemic therapies, and contained grading system in which the grade I and III were extracted for the meta-analysis. 

 
Figure 1- The workflow of the study. PubMed search: 316 papers were found using two keywords of “breast cancer”, and Gene expression and this number reduced to 16 containing microarray datasets (GSEs) and 4 GSEs with the same platform were chosen for the meta-analysis. Data preprocessing: Normalization using VSN and quantile methods and visualization before and after that using box and PCA plots. Batch effect adjustment was also done using ComBat procedure and the result was visualized using three plots (PCA/MDS/Density plots). Data integration: It was done using Meta-analysis with REM model after Q-Q plot visualization. Visualization: It was done using heatmap for top 50 up/down Meta-DEGs, volcano plots for each and merged datasets, and Venn diagram for showing common DEGs across datasets. Functional analysis: it was done by different methods shown in the workflow. 

2-2- Data pre-processing
To carry out the meta-analysis for obtaining the differentially expressed genes, the 4 datasets were downloaded from GEO (GEOquery (Davis and Meltzer, 2007)), and mainly analyzed by the R/Bioconductor package of “GeneMeta” which is based on Z score and FDR (Lusa et al., 2022). The absolute values of z score cutoff was chosen to be greater than 1 with FDR < 0.05. 
• Matching genes between platforms
As we used the same platforms, there were the same number of genes (21,156) in all datasets and there was not a need to identify the common GenBank identifier across the different platforms. 
• Data normalization and visualization using Box and PCA plots
In this meta-analysis, before merging the microarray datasets by the REM method, in order to examine the quality of the datasets to determine whether further data normalization is necessary, datasets were visualized using box and PCA plots before and after normalization. After the first step of visualization, the expression intensities in the datasets were normalized by VSN method (Huber et al., 2002) followed by quantile normalization (Lin et al., 2008). This method is also recommended by the NetworkAnalyst platform for microarray data combination (Zhou et al., 2019). Then, the normalized datasets were visualized by the box and PCA plots.
• Batch effect adjustment by the “ComBat” method and 3 visualization methods (PCA, MDS, and density plots) before and after the adjustment 
In addition to biological factors, technical sources of variations e.g., the use of different platforms or different processing times, which are known as non-biological factors or batch effects, can change the gene expression measurements (Leek et al., 2010). In this study, datasets were corrected for batch effects before applying REM using the ComBat method, from the R/Bioconductor software package “sva” (Leek et al., 2012) used to remove batch effects (Johnson et al., 2006; Jha et al., 2016). Moreover, three visualization methods (PCA, MDS, and density plots) were used before and after the study batch effect correction to see whether datasets are eligible for being merged in the meta-analysis. Multidimensional scaling (MDS) is a means of visualizing the level of similarity among individual datasets which indicates the distance between samples (Mead, 1992) while PCA is based on angles among vectors. 


2-3- Dataset integration 
In this study, for selection of significantly up- and down-regulated genes across four datasets, the random effects model (REM) was applied for combining datasets and a list of the meta combined DEGs (4212) was obtained (|Z score|>1 & FDR< 0.05). Additionally, a common list of 99 DEGs which shared among all 4 datasets was extracted and showed in the result section. Moreover, as the GSE25065 dataset showed dramatically the lowest number of DEGs, we also repeated the meta-analysis after its exclusion to make a comparison between the meta-analysis with 3 and 4 datasets. However, in order to keep the number of sample size high and consequently having more statistical power, the other parts of the analysis i.e., functional analyses, were done using the result of four datasets. Nevertheless, we did some primary comparisons between results of DEGs of both cases (3 or 4 datasets).
• Choosing the best meta-analysis method (among Fishers, REM, FEM)
There are different approaches for combining microarray datasets. Some are based on combining p-values from multiple studies among which Fisher’s and Stouffer’s methods are two of the most well-known approaches. Second ones combine studies using effect sizes and REM and FEM are two of them. To determine which one is appropriate for a meta-analysis, it depends on the level of heterogeneity of individual datasets that has been a challenge of integrating microarray data and assessed by the Cochran’s Q test. When the estimated Q values have an approximately chi-squared distribution, the FEM model is appropriate and otherwise, REM should be used.
In this study, this test was used to examine the homogeneity of the datasets (West et al., 2010) and to assess the consistency among the datasets. Based on this test, the Quantile- Quantile plot was constructed which showed the deviation of the Q values from the chi-squared distribution, and consequently, the REM was chosen as the appropriate model to analyze the included datasets in the meta-analysis. 
For conducting meta-analysis based on the REM model, the “GeneMeta” package in R was implemented which is based on choi et al (Choi et al., 2003). In this package, using implementation of the function “zScores”, the Z scores for each single experiment, and the combined experiments are computed (Lusa et al., 2022). The function “zScores” in the “GeneMeta” package which is based on the approach of Choi et al (Choi et al., 2003) works for a set of expression sets. For each single experiment as well as the combined experiment, the “GeneMeta” package computes the FDR for each gene using the function “zScoreFDR” based on the approach of Choi et al (Choi et al., 2003). In fact, the combined Z score of each gene across datasets is not obtained by simply calculating an average of Z scores of that gene in all datasets. Instead, as shown in the following formula (Equation 1), Z score for each gene (within a single experiment) is calculated by subtracting the mean intensity of all of the genes from the raw intensity data for each gene and dividing that result by the SD of all of the measured intensities. Z score transformation is used to standardize data across a wide range of experiments which regardless of the original hybridization intensities, enables us to compare the microarray data (Cheadle et al., 2003).
Equation 1- Z score formula (Cheadle et al., 2003).
Z score = (intensity G – mean intensity G1…Gn)) / SDG1…Gn

* G is any gene on the microarray & G1…Gn represents the aggregate measure of all of the genes.

• When multiple probes are mapped to the same gene symbol
The microarrays used in this study are oligonucleotide and each probe corresponds to one or a set of short oligonucleotide sequences and there can be multiple probe-sets per each gene name (Raddatz et al., 2017). Therefore, after obtaining DEGs, for each of some probe IDs there was only one gene symbol, but for those genes that matched with multiple probes with different expression measurements, the average values of multiple probes for the same gene were calculated as the expression value of that gene. For this purpose, as there were the same gene names for multiple probe names, probe IDs were removed, and gene symbols were used. 7718 DEGs were reduced to 4212 DEGs after getting the average of multiple probes. This step was done after combining datasets by the REM model in the meta-analysis. 
• Venn diagram
After merging datasets through meta-analysis and removing duplicates, the Venn diagram was produced using the “ggvenn” Package in R for DEGs of each individual study. It also depicts the number of genes that are shared between the existing datasets. The shared number of DEGs across all datasets was reported as “common DEGs”.
• Heatmaps
Two different types of heatmaps were constructed which are shown in functional analysis section. 
• Volcano plots 
For each individual dataset, a volcano plot was constructed using some packages in R including “ggplot2”, “reshape2”, “plotly. Finally, after combining 4 datasets, a combined volcano plot was also produced (Adjusted p-value <0.05 and |z score|>1). 
2-5- Functional analysis
Functional analysis was used to identify the underlying biological processes and mechanisms of the list of DEGs between two phenotypes of the grades of I and III in breast cancer patients. After the meta-combined DEG list (4212 DEGs) was obtained by merging four datasets by REM model in the meta-analysis, this list was used for this purpose. In functional analysis section, through the PPI network construction in the NetworkAnalyst (Zhou et al., 2019), hub genes were identified and the relevant KEGG (Ogata et al., 1999) and GO (BP, MF, CC) pathways for both up- and down DEGs were obtained. Next, two heatmaps for the top 50 up/down DEGs were constructed, and their pathways were identified in this web tool. Moreover, a gene expression heatmap for meta-DEGs was created using the “ComplexHeatmap” R package (Gu et al., 2016). Additionally, the ExpressAnalyst from NetworkAnalyst was used for ORA analysis based on both KEGG and GO (BP, MF, CC) databases. Finally, GSEA analysis was implemented using “fgsea” (Korotkevich et al., 2019) and “msigdbr” (Subramanian et al., 2005) packages in R. 
2-5-1- Hub genes, KEGG and GO (BP, MF, CC) pathways for the PPI network
To find out the hub genes among the DEGs (4212) from the meta-analysis of 4 datasets, the PPI network was constructed using the NetworkAnalyst. The gene list was first imported through the “One Gene List” input in the NetworkAnalyst. For the PPI network construction, the tissue specific PPI (breast mammary tissue) database was chosen that have been collected from the DifferentialNet database (Basha et al., 2018). When tissue specific PPI is applied, false positives can dramatically be reduced (Zhou et al., 2019). In order to avoid the hairball effect and visualize in a better way, the “Zero-order” network was applied. When there is a large number of query genes (‘seeds’), Zero-order network method is recommended (Zhou et al., 2019). After implementation of PPI network, hub genes containing the number of degrees were obtained. The degree of a node indicates the number of connections that has with other nodes (Zhou et al., 2019). Then, using the KEGG and GO databases, the most significant pathways were identified. Finally, using the “module explorer” panel which is based on the WalkTrap algorithm (Pons et al., 2005), the most important modules of up- and down- regulated DEGs were extracted.
2-5-2- ORA
The ExpressAnalyst was also employed for implementation of ORA method to visualize functional categories enriched in a network. Over Representation Analysis was carried out to determine whether known biological functions or processes are over-represented in the meta- DEG list of four datasets (Boyle et al., 2004). For this analysis, KEGG and GO (BP, MF, CC) databases were applied and top 5 sig pathways for each of them were extracted. 
2-5-3- Heatmaps
Two different types of heatmaps were constructed as follows
• Similarity matrix and gene expression heatmap
Using the “ComplexHeatmap” R package in Bioconductor, after merging datasets by REM model in the meta-analysis, two combined heatmaps for top 50 up/down- regulated DEGs across 4 and 3 datasets were produced. All patient samples with the histologic grade I and III tumors were aggregated separately on the x axis which presents changes between these two grades across top 50 up/down DEGs more easily. 
• Heatmaps for top 50 meta-combined DEGs
Using the ExpressAnalyst from the NetworkAnalyst web tool, two other heatmaps were constructed for the top 50 up & top 50 down DEGs from the list of DEGs (4212) (the 4 dataset meta-analysis). Then, the relevant KEGG pathways for selected genes in heatmaps were identified.
2-5-4- Gene Set Enrichment Analysis (GSEA)
GSEA was also implemented on the meta-DEG list of four datasets (4212 DEGs) using the R/Bioconductor package of "fgsea". To extract the biological insight and improve the interpretability the gene list, GSEA analysis was used which is based on grouping of genes into gene sets that share common biological function, chromosomal location, or regulation (Subramanian et al., 2005).

		</p>
	</div>

	
	
	<div class="content">
		<h2 class="heading">Content 2</h2>
		<p>
			  TEEEEXXXXTTTTT BEFORE TABLE1

			<a href="table1.html">(table 1)</a>,

			  TEEEEXXXXTTTTT FIGURE1

			<a href="figure1.html">(figure 1)</a>. 

			  - Introduction
• The importance of Breast cancer among other cancers and diseases  
In 2020, cancers (602,350 deaths), followed by heart disease (696,962 deaths), was the second biggest cause of deaths in the United States, according to the Centers for Disease Control and Prevention (CDC) (CDC, 2022). The International Agency for Research on Cancer also published the GLOBOCAN 2020 estimates of cancer incidence and mortality, worldwide for 36 cancers in 185 countries. They estimated that 19.3 million new cancer cases and almost 10.0 million cancer deaths occurred in 2020. Female breast cancer with an estimated 2.3 million new cases (11.7%) was the most commonly diagnosed cancer, followed by lung (11.4%), colorectal (10.0 %), prostate (7.3%), and stomach (5.6%) cancers. In terms of mortality, female breast cancer, with 685,000 deaths, obtained the fifth leading cause of cancer death (6.9%), and lung (18%), colorectal (9.4%), liver (8.3%), stomach (7.7%) cancers were the first four leading causes of cancer death worldwide (Sung et al., 2021). Additionally, in 2018, breast cancer in women was the second most common cause of cancer death in the UK (11,500 deaths) (Cancer Research UK, 2015).

 • Metastasis in breast cancer and importance of early detection by potential new biomarkers 
Approximately, 10% of patients with early-stage breast cancer may develop metastatic breast cancer (Colzani et al., 2014). Breast cancer is a heterogeneous disease with dysregulated genes and pathways and different responses to treatments which makes it difficult to manage, especially in aggressive subtypes. However, the “aggressiveness” is not a standard term for breast cancer and there is a need to identify factors for the purpose of screening, prognosis, and treatment responses in the management of the disease (Arpino et al., 2015). Furthermore, early detection of breast cancer using early diagnosis and screening methods is vital for those with more advanced breast cancer subtypes that can result in early treatments of detected patients, and reduced cancer mortality as well (WHO, 2022). Biomarker identification in breast cancer can help develop more accurate prognosis and efficient treatments as different treatment strategies are needed to deal with metastatic breast cancers (MBCs) compared to localized ones (Kim, 2021).

• The importance of gene expression profiling for early-stage detection of metastatic patients
To Identify biomarkers with potential diagnostic and therapeutic targets in breast cancer patients, gene expression profiling has been used (Fu et al., 2014). It can be also used to predict aggressive and metastatic breast cancer patients and identification of candidate targets for new drugs (Colzani et al., 2014). Gene expression tests, also called gene expression profiling, are done on breast cancer cells after surgery or biopsy to identify the patterns of a number of different genes. When breast cancer is advanced, even without these tests, it might be clear that chemotherapy is needed. Therefore, these tests, such as the Oncotype DX, MammaPrint, and Prosigna, are used for early-stage cancers to look at different sets of breast cancer genes and see whether chemotherapy is needed to reduce the risk of cancer recurrence (American Cancer Society, 2022). Expression profiling has so far been used to detect breast cancer women with a poor prognosis who can benefit from adjuvant therapy (Pawitan et al., 2005). Pathway signatures, on the other hand, is used to discover biomarkers for drug predictions. Evidence shows that Identifying proteins upstream of a pathway that are responsible in the growth of the tumor cells and delivering anti-cancer drugs to appropriate patients can have inhibitory effects (Itadani et al., 2008). Recurrence or progression of a disease can also be checked in diagnosed patients using a prognostic marker (Pepe et al., 2008). Microarray and RNA-sequencing technologies as a means of gene expression profiling which provides prognostic and predictive approaches, widely used in oncology and cancer management (Supplitt et al., 2021). 
Moreover, histological tumor grade is one of numerous prognostic factors in early-stage breast cancer (Cianfrocca, 2004) which can be used to determine whether a breast cancer is metastatic or to choose the appropriate treatment for individual patients (Arpino et al., 2015). Tumor grade is a prognostic marker in early-stage breast cancer. One study showed that hazard ratio for grade III vs grade I was 1.95 which indicates a significant difference between these two grades of breast cancer (Poikonen et al., 2005). In our study, therefore, using microarray datasets from GEO database, two distinct groups of grades I and III of breast cancer tumors were compared together to reveal significant changes in DEGs and pathways and to obtain a new gene expression pattern and introduction of some candidate genes as biomarkers. 

• Integration and meta-analysis of microarray datasets & the increase in the statistical power
DNA microarrays have been used to study patterns of gene expression between healthy and diseased cells and tissues (Fumagalli et al., 2014) and discovering differentially expressed genes (Kim et al., 2007). One of the main limitations of microarray datasets is the small number of sample sizes of individual studies which leads to low statistical power, but with the integration of independent datasets, the statistical power is also increased (Kim et al., 2007). Horizontal and vertical data integrations are two general types to combine individual studies (Tseng et al., 2012). Horizontal data integration, which is widely used in meta-analysis, involves the combination of different datasets with the same data type such transcriptomic data, while vertical data integration involves multi-omics data Integration e.g., combining metabolic and transcriptomic datasets from the same patient group (Xia et al., 2013).  Over the past decade, horizontal integration and meta-analysis of microarray datasets has been broadly used in transcriptomic research (Tseng et al., 2012). 

• Meta-analysis importance and the existing methods
With the increase in the microarray data in open-source databases, meta-analysis and data integration methods have become vital to combine results from different studies with small number of sample sizes to considerably increase the sample sizes and statistical power, consequently (Choi et al., 2004; Marot et al., 2009). Meta-analysis has also been widely used in medicine and health policy to interpret contradictory results from various studies (Haidich, 2010). 
A broad range of meta-analysis methods have so far been introduced. There are two major categories including combined p-values (e.g., Fisher’s and Stouffer’s methods) and combined effect sizes (e.g., fixed & random effects models) (Chang et al., 2013). Rhodes et al. (2002) and Choi et al. (2003) were amongst the first authors to establish and demonstrate the meta-analysis and systematic integration of multiple independent microarray datasets to find differentially expressed (DE) genes (Marot et al., 2009). Choi et al. introduced a method based on the effect size calculated by combinations of expression values, whereas Rhodes et al. used the method in which p-values are combined (Rhodes et al., 2002; Choi et al., 2003). Conlon et al. (2007) proposed the Bayesian meta-analysis model to combine microarray data which is the probability integration model. In their model, by combining probabilities across studies rather than aggregating the gene expression measures (which accounts for variability across studies), they avoided an inter-study variability parameter. Therefore, they believed that this model for their datasets identified more true discovered genes and fewer true omitted genes than combining expression measures (Conlon et al., 2007).
In this study, using the horizontal integration, we carried out a microarray meta-analysis (Random Effects Model known as REM) to combine microarray datasets with the same platform from GEO database. This meta-analysis was done with the purpose of obtaining a new gene expression profile for more aggressive BC patients, with analyzing data from breast cancer patients with the histologic grade I tumors against grade III ones. Then, we obtained a list of meta-combined DEGs as well as a common DEG list which shared across all individual studies that might be used as a gene expression pattern to distinguish between a group of patients with a high probability of progression from early stage to metastasis. Finally, the functional analysis (such as PPI network construction, hub genes, KEGG pathways and GO pathways identification, and etc.) was employed to identify the underlying biological processes and mechanisms of the obtained meta-combined DEG list between two phenotypes of the grades of I and III in breast cancer patients. 

		</p>
	</div>


	<div class="content">
		<h2 class="heading">Content 3</h2>
		<ul>
			<li>Subheading 1</li>
			<li>Subheading 2</li>
			<li>Subheading 3</li>
			<li>Subheading 4</li>
		</ul>
	</div>

	<div class="content">
		<h2 class="heading">Content 4</h2>
		<ul>
			<li>Subheading 1</li>
			<li>Subheading 2</li>
			<li>Subheading 3</li>
			<li>Subheading 4</li>
		</ul>
	</div>

	<div class="content">
		<h2 class="heading">Content 5</h2>
		<ul>
			<li>Subheading 1</li>
			<li>Subheading 2</li>
			<li>Subheading 3</li>
			<li>Subheading 4</li>
		</ul>
	</div>

	<div class="content">
		<h2 class="heading">Content 6</h2>
		<ul>
			<li>Subheading 1</li>
			<li>Subheading 2</li>
			<li>Subheading 3</li>
			<li>Subheading 4</li>
		</ul>
	</div>

	<div class="content">
		<h2 class="heading">Content 7</h2>
		<ul>
			<li>Subheading 1</li>
			<li>Subheading 2</li>
			<li>Subheading 3</li>
			<li>Subheading 4</li>
		</ul>
	</div>

	<script>
		  // Get all the headings
		  const headings = document.querySelectorAll(".heading");

		  // Add event listener to each heading
		  headings.forEach((heading) => {
		    heading.addEventListener("click", () => {
		      const subheadings = heading.nextElementSibling;
		      subheadings.style.display = subheadings.style.display === "none" ? "block" : "none";
		      heading.classList.toggle("active"); // Toggle the active class
		    });
		  });

		  // Add event listener to table 1 link
		  const table1Link = document.querySelector('a[href="table1.html"]');
		  table1Link.addEventListener("click", () => {
		    window.open("table1.html", "_blank");
		  });

		  // Add event listener to figure 1 link
		  const figure1Link = document.querySelector('a[href="figure1.html"]');
		  figure1Link.addEventListener("click", () => {
		    window.open("figure1.html", "_blank");
		  });
	</script>

</body>
</html>


		
		
		
		
